<!DOCTYPE HTML>
<html lang="zh-CN">


<head>
    <meta charset="utf-8">
    <meta name="keywords" content="HIERARCHICAL CLUSTERING IN R (层级聚类及其在R中实现), Shaobo Liang">
    <meta name="description" content="HIERARCHICAL CLUSTERING IN R: THE ESSENTIALS (层级聚类)link1-lessonlink2-heatmap-extend
Agglomerative Hierarchical Clusterin">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=no">
    <meta name="renderer" content="webkit|ie-stand|ie-comp">
    <meta name="mobile-web-app-capable" content="yes">
    <meta name="format-detection" content="telephone=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    <meta name="referrer" content="no-referrer-when-downgrade">
    <!-- Global site tag (gtag.js) - Google Analytics -->


    <title>HIERARCHICAL CLUSTERING IN R (层级聚类及其在R中实现) | Shaobo Liang</title>
    <link rel="icon" type="image/png" href="/favicon.png">

    <link rel="stylesheet" type="text/css" href="/libs/awesome/css/all.min.css">
    <link rel="stylesheet" type="text/css" href="/libs/materialize/materialize.min.css">
    <link rel="stylesheet" type="text/css" href="/libs/aos/aos.css">
    <link rel="stylesheet" type="text/css" href="/libs/animate/animate.min.css">
    <link rel="stylesheet" type="text/css" href="/libs/lightGallery/css/lightgallery.min.css">
    <link rel="stylesheet" type="text/css" href="/css/matery.css">
    <link rel="stylesheet" type="text/css" href="/css/my.css">

    <script src="/libs/jquery/jquery-3.6.0.min.js"></script>

<meta name="generator" content="Hexo 6.0.0">
<style>.github-emoji { position: relative; display: inline-block; width: 1.2em; min-height: 1.2em; overflow: hidden; vertical-align: top; color: transparent; }  .github-emoji > span { position: relative; z-index: 10; }  .github-emoji img, .github-emoji .fancybox { margin: 0 !important; padding: 0 !important; border: none !important; outline: none !important; text-decoration: none !important; user-select: none !important; cursor: auto !important; }  .github-emoji img { height: 1.2em !important; width: 1.2em !important; position: absolute !important; left: 50% !important; top: 50% !important; transform: translate(-50%, -50%) !important; user-select: none !important; cursor: auto !important; } .github-emoji-fallback { color: inherit; } .github-emoji-fallback img { opacity: 0 !important; }</style>
</head>



   <style>
    body{
       background-image: url(http://h2.ioliu.cn/bing/BulgariaPerseids_EN-AU11585904087_1920x1080.jpg?imageslim&quot);
       background-repeat:no-repeat;
       background-size: 100% 100%;
       background-attachment:fixed;
    }
</style>



<body>
    <header class="navbar-fixed">
    <nav id="headNav" class="bg-color nav-transparent">
        <div id="navContainer" class="nav-wrapper container">
            <div class="brand-logo">
                <a href="/" class="waves-effect waves-light">
                    
                    <img src="/medias/logo.png" class="logo-img" alt="LOGO">
                    
                    <span class="logo-span">Shaobo Liang</span>
                </a>
            </div>
            

<a href="#" data-target="mobile-nav" class="sidenav-trigger button-collapse"><i class="fas fa-bars"></i></a>
<ul class="right nav-menu">
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/" class="waves-effect waves-light">
      
      <i class="fas fa-home" style="zoom: 0.6;"></i>
      
      <span>首页</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/tags" class="waves-effect waves-light">
      
      <i class="fas fa-tags" style="zoom: 0.6;"></i>
      
      <span>标签</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/categories" class="waves-effect waves-light">
      
      <i class="fas fa-bookmark" style="zoom: 0.6;"></i>
      
      <span>分类</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/archives" class="waves-effect waves-light">
      
      <i class="fas fa-archive" style="zoom: 0.6;"></i>
      
      <span>归档</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/about" class="waves-effect waves-light">
      
      <i class="fas fa-user-circle" style="zoom: 0.6;"></i>
      
      <span>关于</span>
    </a>
    
  </li>
  
  <li>
    <a href="#searchModal" class="modal-trigger waves-effect waves-light">
      <i id="searchIcon" class="fas fa-search" title="搜索" style="zoom: 0.85;"></i>
    </a>
  </li>
</ul>


<div id="mobile-nav" class="side-nav sidenav">

    <div class="mobile-head bg-color">
        
        <img src="/medias/logo.png" class="logo-img circle responsive-img">
        
        <div class="logo-name">Shaobo Liang</div>
        <div class="logo-desc">
            
            成长之中，wait for me~
            
        </div>
    </div>

    <ul class="menu-list mobile-menu-list">
        
        <li class="m-nav-item">
	  
		<a href="/" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-home"></i>
			
			首页
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="/tags" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-tags"></i>
			
			标签
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="/categories" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-bookmark"></i>
			
			分类
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="/archives" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-archive"></i>
			
			归档
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="/about" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-user-circle"></i>
			
			关于
		</a>
          
        </li>
        
        
        <li><div class="divider"></div></li>
        <li>
            <a href="https://github.com/blinkfox/hexo-theme-matery" class="waves-effect waves-light" target="_blank">
                <i class="fab fa-github-square fa-fw"></i>Fork Me
            </a>
        </li>
        
    </ul>
</div>


        </div>

        
            <style>
    .nav-transparent .github-corner {
        display: none !important;
    }

    .github-corner {
        position: absolute;
        z-index: 10;
        top: 0;
        right: 0;
        border: 0;
        transform: scale(1.1);
    }

    .github-corner svg {
        color: #0f9d58;
        fill: #fff;
        height: 64px;
        width: 64px;
    }

    .github-corner:hover .octo-arm {
        animation: a 0.56s ease-in-out;
    }

    .github-corner .octo-arm {
        animation: none;
    }

    @keyframes a {
        0%,
        to {
            transform: rotate(0);
        }
        20%,
        60% {
            transform: rotate(-25deg);
        }
        40%,
        80% {
            transform: rotate(10deg);
        }
    }
</style>

<a href="https://github.com/blinkfox/hexo-theme-matery" class="github-corner tooltipped hide-on-med-and-down" target="_blank"
   data-tooltip="Fork Me" data-position="left" data-delay="50">
    <svg viewBox="0 0 250 250" aria-hidden="true">
        <path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path>
        <path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2"
              fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path>
        <path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z"
              fill="currentColor" class="octo-body"></path>
    </svg>
</a>
        
    </nav>

</header>

    



<div class="bg-cover pd-header post-cover" style="background-image: url('/medias/featureimages/23.jpg')">
    <div class="container" style="right: 0px;left: 0px;">
        <div class="row">
            <div class="col s12 m12 l12">
                <div class="brand">
                    <h1 class="description center-align post-title">HIERARCHICAL CLUSTERING IN R (层级聚类及其在R中实现)</h1>
                </div>
            </div>
        </div>
    </div>
</div>




<main class="post-container content">

    
    <link rel="stylesheet" href="/libs/tocbot/tocbot.css">
<style>
    #articleContent h1::before,
    #articleContent h2::before,
    #articleContent h3::before,
    #articleContent h4::before,
    #articleContent h5::before,
    #articleContent h6::before {
        display: block;
        content: " ";
        height: 100px;
        margin-top: -100px;
        visibility: hidden;
    }

    #articleContent :focus {
        outline: none;
    }

    .toc-fixed {
        position: fixed;
        top: 64px;
    }

    .toc-widget {
        width: 345px;
        padding-left: 20px;
    }

    .toc-widget .toc-title {
        padding: 35px 0 15px 17px;
        font-size: 1.5rem;
        font-weight: bold;
        line-height: 1.5rem;
    }

    .toc-widget ol {
        padding: 0;
        list-style: none;
    }

    #toc-content {
        padding-bottom: 30px;
        overflow: auto;
    }

    #toc-content ol {
        padding-left: 10px;
    }

    #toc-content ol li {
        padding-left: 10px;
    }

    #toc-content .toc-link:hover {
        color: #42b983;
        font-weight: 700;
        text-decoration: underline;
    }

    #toc-content .toc-link::before {
        background-color: transparent;
        max-height: 25px;

        position: absolute;
        right: 23.5vw;
        display: block;
    }

    #toc-content .is-active-link {
        color: #42b983;
    }

    #floating-toc-btn {
        position: fixed;
        right: 15px;
        bottom: 76px;
        padding-top: 15px;
        margin-bottom: 0;
        z-index: 998;
    }

    #floating-toc-btn .btn-floating {
        width: 48px;
        height: 48px;
    }

    #floating-toc-btn .btn-floating i {
        line-height: 48px;
        font-size: 1.4rem;
    }
</style>
<div class="row">
    <div id="main-content" class="col s12 m12 l9">
        <!-- 文章内容详情 -->
<div id="artDetail">
    <div class="card">
        <div class="card-content article-info">
            <div class="row tag-cate">
                <div class="col s7">
                    
                    <div class="article-tag">
                        
                            <a href="/tags/hierarchical-clustering/">
                                <span class="chip bg-color">hierarchical clustering</span>
                            </a>
                        
                            <a href="/tags/R/">
                                <span class="chip bg-color">R</span>
                            </a>
                        
                    </div>
                    
                </div>
                <div class="col s5 right-align">
                    
                </div>
            </div>

            <div class="post-info">
                
                <div class="post-date info-break-policy">
                    <i class="far fa-calendar-minus fa-fw"></i>发布日期:&nbsp;&nbsp;
                    2021-12-30
                </div>
                

                
                <div class="post-date info-break-policy">
                    <i class="far fa-calendar-check fa-fw"></i>更新日期:&nbsp;&nbsp;
                    2021-12-30
                </div>
                

                

                

                
            </div>
        </div>
        <hr class="clearfix">

        

        

        <div class="card-content article-card-content">
            <div id="articleContent">
                <h1 id="HIERARCHICAL-CLUSTERING-IN-R-THE-ESSENTIALS-层级聚类"><a href="#HIERARCHICAL-CLUSTERING-IN-R-THE-ESSENTIALS-层级聚类" class="headerlink" title="HIERARCHICAL CLUSTERING IN R: THE ESSENTIALS (层级聚类)"></a>HIERARCHICAL CLUSTERING IN R: THE ESSENTIALS (层级聚类)</h1><p><a target="_blank" rel="noopener" href="https://www.datanovia.com/en/lessons/agglomerative-hierarchical-clustering/">link1-lesson</a><br><a target="_blank" rel="noopener" href="https://www.datanovia.com/en/blog/clustering-using-correlation-as-distance-measures-in-r/">link2-heatmap-extend</a></p>
<h2 id="Agglomerative-Hierarchical-Clustering"><a href="#Agglomerative-Hierarchical-Clustering" class="headerlink" title="Agglomerative Hierarchical Clustering"></a>Agglomerative Hierarchical Clustering</h2><p><a target="_blank" rel="noopener" href="https://www.datanovia.com/en/lessons/agglomerative-hierarchical-clustering/">丛集聚类</a><br>分裂聚类</p>
<img src="hierarchical_clustering.jpg" width="600">


<h2 id="Steps-to-agglomerative-hierarchical-clustering"><a href="#Steps-to-agglomerative-hierarchical-clustering" class="headerlink" title="Steps to agglomerative hierarchical clustering"></a>Steps to agglomerative hierarchical clustering</h2><center><img src="Steps_to_agglomerative.jpg" width="800"></center>

<h3 id="Data-structure-and-preparation"><a href="#Data-structure-and-preparation" class="headerlink" title="Data structure and preparation"></a>Data structure and preparation</h3><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Load the data</span></span><br><span class="line">data(<span class="string">"USArrests"</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Standardize the data</span></span><br><span class="line"><span class="comment">## The R function scale() can be used for standardization</span></span><br><span class="line">df &lt;- scale(USArrests)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Show the first 6 rows</span></span><br><span class="line">head(df, nrow = <span class="number">6</span>)</span><br></pre></td></tr></tbody></table></figure>


<table class="dataframe">
<caption>A matrix: 6 × 4 of type dbl</caption>
<thead>
    <tr><th></th><th scope="col">Murder</th><th scope="col">Assault</th><th scope="col">UrbanPop</th><th scope="col">Rape</th></tr>
</thead>
<tbody>
    <tr><th scope="row">Alabama</th><td>1.24256408</td><td>0.7828393</td><td>-0.5209066</td><td>-0.003416473</td></tr>
    <tr><th scope="row">Alaska</th><td>0.50786248</td><td>1.1068225</td><td>-1.2117642</td><td> 2.484202941</td></tr>
    <tr><th scope="row">Arizona</th><td>0.07163341</td><td>1.4788032</td><td> 0.9989801</td><td> 1.042878388</td></tr>
    <tr><th scope="row">Arkansas</th><td>0.23234938</td><td>0.2308680</td><td>-1.0735927</td><td>-0.184916602</td></tr>
    <tr><th scope="row">California</th><td>0.27826823</td><td>1.2628144</td><td> 1.7589234</td><td> 2.067820292</td></tr>
    <tr><th scope="row">Colorado</th><td>0.02571456</td><td>0.3988593</td><td> 0.8608085</td><td> 1.864967207</td></tr>
</tbody>
</table>



<h3 id="Similarity-measures"><a href="#Similarity-measures" class="headerlink" title="Similarity measures"></a>Similarity measures</h3><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Compute the dissimilarity matrix</span></span><br><span class="line"><span class="comment"># df = the standardized data</span></span><br><span class="line">res.dist &lt;- dist(df, method = <span class="string">"euclidean"</span>)  <span class="comment"># Euclidean and manhattan</span></span><br></pre></td></tr></tbody></table></figure>


<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># view distance</span></span><br><span class="line">as.matrix(res.dist)[<span class="number">1</span>:<span class="number">6</span>, <span class="number">1</span>:<span class="number">6</span>]</span><br></pre></td></tr></tbody></table></figure>


<table class="dataframe">
<caption>A matrix: 6 × 6 of type dbl</caption>
<thead>
    <tr><th></th><th scope="col">Alabama</th><th scope="col">Alaska</th><th scope="col">Arizona</th><th scope="col">Arkansas</th><th scope="col">California</th><th scope="col">Colorado</th></tr>
</thead>
<tbody>
    <tr><th scope="row">Alabama</th><td>0.000000</td><td>2.703754</td><td>2.293520</td><td>1.289810</td><td>3.263110</td><td>2.651067</td></tr>
    <tr><th scope="row">Alaska</th><td>2.703754</td><td>0.000000</td><td>2.700643</td><td>2.826039</td><td>3.012541</td><td>2.326519</td></tr>
    <tr><th scope="row">Arizona</th><td>2.293520</td><td>2.700643</td><td>0.000000</td><td>2.717758</td><td>1.310484</td><td>1.365031</td></tr>
    <tr><th scope="row">Arkansas</th><td>1.289810</td><td>2.826039</td><td>2.717758</td><td>0.000000</td><td>3.763641</td><td>2.831051</td></tr>
    <tr><th scope="row">California</th><td>3.263110</td><td>3.012541</td><td>1.310484</td><td>3.763641</td><td>0.000000</td><td>1.287619</td></tr>
    <tr><th scope="row">Colorado</th><td>2.651067</td><td>2.326519</td><td>1.365031</td><td>2.831051</td><td>1.287619</td><td>0.000000</td></tr>
</tbody>
</table>



<h3 id="Linkage"><a href="#Linkage" class="headerlink" title="Linkage"></a>Linkage</h3><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">res.hc &lt;- hclust(d = res.dist, method = <span class="string">"ward.D2"</span>)</span><br></pre></td></tr></tbody></table></figure>

<h3 id="Verify-the-cluster-tree"><a href="#Verify-the-cluster-tree" class="headerlink" title="Verify the cluster tree"></a>Verify the cluster tree</h3><p>One way to measure how well the cluster tree generated by the hclust() function reflects your data is to compute the correlation between the cophenetic distances and the original distance data generated by the dist() function.<br>If the clustering is valid, the linking of objects in the cluster tree should have a strong correlation with the distances between objects in the original distance matrix.<br>评估聚类效果的方法之一，就是评估 <code>hclust()</code> 算出来的距离 和 用<code>dist()</code>算的原始数据的距离，二者之间的相关性好不好。</p>
<p>The closer the value of the correlation coefficient is to 1, the more accurately the clustering solution reflects your data. Values above 0.75 are felt to be good.<br>The “average” linkage method appears to produce high values of this statistic.<br>相关性越接近 <strong>1</strong>，说明聚类的效果越能反映真实情况.</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Compute cophentic distance</span></span><br><span class="line">res.coph &lt;- cophenetic(res.hc)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Correlation between cophenetic distance and</span></span><br><span class="line"><span class="comment"># the original distance</span></span><br><span class="line">cor(res.dist, res.coph)</span><br></pre></td></tr></tbody></table></figure>


<p>0.697526563237039</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Execute the hclust() function again using the average linkage method</span></span><br><span class="line">res.hc2 &lt;- hclust(res.dist, method = <span class="string">"average"</span>)</span><br><span class="line"></span><br><span class="line">cor(res.dist, cophenetic(res.hc2))</span><br></pre></td></tr></tbody></table></figure>


<p>0.718038237932047</p>
<h3 id="Dendrogram-树状图"><a href="#Dendrogram-树状图" class="headerlink" title="Dendrogram(树状图)"></a>Dendrogram(树状图)</h3><p>Dendrogram can be produced in R using the base function <code>plot(res.hc)</code>, where res.hc is the output of hclust()</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plot(res.hc, cex = <span class="number">0.8</span>)</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_18_0.png" alt="png"></p>
<p> <code>fviz_dend()</code> ( in <code>factoextra</code> R package ) to produce a beautiful dendrogram.</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># cex: label size</span></span><br><span class="line">suppressPackageStartupMessages(library(<span class="string">"factoextra"</span>))</span><br><span class="line">fviz_dend(res.hc, title=<span class="string">''</span>, cex = <span class="number">0.7</span>)</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_20_0.png" alt="png"></p>
<h3 id="Cut-the-dendrogram-into-different-groups"><a href="#Cut-the-dendrogram-into-different-groups" class="headerlink" title="Cut the dendrogram into different groups"></a>Cut the dendrogram into different groups</h3><p>You can cut the hierarchical tree at a given height in order to partition your data into clusters.<br>The R base function cutree() can be used to cut a tree, generated by the hclust() function, into several groups either by specifying the desired number of groups or the cut height.</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Cut tree into 4 groups</span></span><br><span class="line">grp &lt;- cutree(res.hc, k = <span class="number">4</span>)</span><br><span class="line">head(grp, n = <span class="number">4</span>)</span><br></pre></td></tr></tbody></table></figure>


<style>
.dl-inline {width: auto; margin:0; padding: 0}
.dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}
.dl-inline>dt::after {content: ":\0020"; padding-right: .5ex}
.dl-inline>dt:not(:first-of-type) {padding-left: .5ex}
</style><dl class="dl-inline"><dt>Alabama</dt><dd>1</dd><dt>Alaska</dt><dd>2</dd><dt>Arizona</dt><dd>2</dd><dt>Arkansas</dt><dd>3</dd></dl>




<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Number of members in each cluster</span></span><br><span class="line">table(grp)</span><br></pre></td></tr></tbody></table></figure>


<pre><code>grp
 1  2  3  4 
 7 12 19 12 
</code></pre>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Get the names for the members of cluster 1</span></span><br><span class="line">rownames(df)[grp == <span class="number">1</span>]</span><br></pre></td></tr></tbody></table></figure>


<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>'Alabama'</li><li>'Georgia'</li><li>'Louisiana'</li><li>'Mississippi'</li><li>'North Carolina'</li><li>'South Carolina'</li><li>'Tennessee'</li></ol>



<p><a target="_blank" rel="noopener" href="http://agroninfotech.blogspot.com/2020/06/visualizing-clusters-in-r-hierarchical.html"><strong>fviz_dend advanced handbook</strong></a></p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">fviz_dend(res.hc, title=<span class="string">''</span>, cex=<span class="number">0.7</span>, k=<span class="number">4</span>, type=<span class="string">'rectangle'</span>, rect=<span class="built_in">T</span> ,rect_border = <span class="string">"jco"</span>, rect_fill = <span class="literal">TRUE</span>, horiz = <span class="literal">TRUE</span>) <span class="comment"># type='scatter'</span></span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_27_0.png" alt="png"></p>
<h2 id="Cluster-R-package"><a href="#Cluster-R-package" class="headerlink" title="Cluster R package"></a>Cluster R package</h2><p>The R package <code>cluster</code> makes it easy to perform cluster analysis in R. It provides the function agnes() and diana() for computing agglomerative and divisive clustering, respectively. These functions perform all the necessary steps for you. You don’t need to execute the scale(), dist() and hclust() function separately.</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">library(<span class="string">"cluster"</span>)</span><br><span class="line"><span class="comment"># Agglomerative Nesting (Hierarchical Clustering)</span></span><br><span class="line">res.agnes &lt;- agnes(x = USArrests, <span class="comment"># data matrix</span></span><br><span class="line">                   stand = <span class="literal">TRUE</span>, <span class="comment"># Standardize the data</span></span><br><span class="line">                   metric = <span class="string">"euclidean"</span>, <span class="comment"># metric for distance matrix</span></span><br><span class="line">                   method = <span class="string">"ward"</span> <span class="comment"># Linkage method</span></span><br><span class="line">                   )</span><br><span class="line"></span><br><span class="line"><span class="comment"># DIvisive ANAlysis Clustering</span></span><br><span class="line">res.diana &lt;- diana(x = USArrests, <span class="comment"># data matrix</span></span><br><span class="line">                   stand = <span class="literal">TRUE</span>, <span class="comment"># standardize the data</span></span><br><span class="line">                   metric = <span class="string">"euclidean"</span> <span class="comment"># metric for distance matrix</span></span><br><span class="line">                   )</span><br></pre></td></tr></tbody></table></figure>

<pre><code>Warning message:
"package 'cluster' was built under R version 4.0.5"
</code></pre>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">fviz_dend(res.diana, cex = <span class="number">0.7</span>, k = <span class="number">4</span>, horiz = <span class="built_in">T</span>, type = <span class="string">'circular'</span>)</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_31_0.png" alt="png"></p>
<h2 id="Application-of-hierarchical-clustering-to-gene-expression-data-analysis"><a href="#Application-of-hierarchical-clustering-to-gene-expression-data-analysis" class="headerlink" title="Application of hierarchical clustering to gene expression data analysis"></a>Application of hierarchical clustering to gene expression data analysis</h2><p>In gene expression data analysis, clustering is generaly used as one of the first step to explore the data.<br>We are interested in whether there are groups of genes or groups of samples that have similar gene expression patterns.  </p>
<p>Several <a target="_blank" rel="noopener" href="https://www.datanovia.com/en/lessons/clustering-distance-measures/">clustering distance measures</a> have been described for assessing the similarity or the dissimilarity between items, in order to decide which items have to be grouped together or not.<br>These measures can be used to cluster genes or samples that are similar.</p>
<p>For most common clustering softwares, the default distance measure is the Euclidean distance.<br>The most popular methods for gene expression data are to use log2(expression + 0.25), correlation distance and complete linkage clustering agglomerative-clustering.  </p>
<p>Note that, when the data are scaled, the Euclidean distance of the z-scores is the same as correlation distance.</p>
<p>Pearson’s correlation is quite sensitive to outliers. When clustering genes, it is important to be aware of the possible impact of outliers.<br>An alternative option is to use Spearman’s correlation instead of Pearson’s correlation.</p>
<h2 id="Comparing-Cluster-Dendrograms-in-R"><a href="#Comparing-Cluster-Dendrograms-in-R" class="headerlink" title="Comparing Cluster Dendrograms in R"></a>Comparing Cluster Dendrograms in R</h2><h3 id="Data-preparation"><a href="#Data-preparation" class="headerlink" title="Data preparation"></a>Data preparation</h3><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">df &lt;- scale(USArrests)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Subset containing 10 rows</span></span><br><span class="line">set.seed(<span class="number">123</span>)</span><br><span class="line">ss &lt;- sample(<span class="number">1</span>:<span class="number">50</span>, <span class="number">10</span>)</span><br><span class="line">df &lt;- df[ss,]</span><br></pre></td></tr></tbody></table></figure>

<h3 id="Dendrograms-comparison"><a href="#Dendrograms-comparison" class="headerlink" title="Dendrograms comparison"></a>Dendrograms comparison</h3><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">suppressPackageStartupMessages(library(dendextend))</span><br><span class="line"></span><br><span class="line"><span class="comment"># Compute distance matrix</span></span><br><span class="line">res.dist &lt;- dist(df, method = <span class="string">"euclidean"</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Compute 2 hierarchical clusterings</span></span><br><span class="line">hc1 &lt;- hclust(res.dist, method = <span class="string">"average"</span>)</span><br><span class="line">hc2 &lt;- hclust(res.dist, method = <span class="string">"ward.D2"</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Create two dendrograms</span></span><br><span class="line">dend1 &lt;- as.dendrogram (hc1)</span><br><span class="line">dend2 &lt;- as.dendrogram (hc2)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Create a list to hold dendrograms</span></span><br><span class="line">dend_list &lt;- dendlist(dend1, dend2)</span><br></pre></td></tr></tbody></table></figure>

<ol>
<li>Visual comparison of two dendrograms</li>
</ol>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Align and plot two dendrograms side by side</span></span><br><span class="line">dendlist(dend1, dend2) %&gt;%</span><br><span class="line">  untangle(method = <span class="string">"step1side"</span>) %&gt;% <span class="comment"># Find the best alignment layout</span></span><br><span class="line">  tanglegram()                       <span class="comment"># Draw the two dendrograms</span></span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_41_0.png" alt="png"></p>
<h2 id="Heatmap-in-R-Static-and-Interactive-Visualization"><a href="#Heatmap-in-R-Static-and-Interactive-Visualization" class="headerlink" title="Heatmap in R: Static and Interactive Visualization"></a>Heatmap in R: Static and Interactive Visualization</h2><h3 id="R-Packages-functions-for-drawing-heatmaps"><a href="#R-Packages-functions-for-drawing-heatmaps" class="headerlink" title="R Packages/functions for drawing heatmaps"></a>R Packages/functions for drawing heatmaps</h3><ul>
<li>heatmap() [R base function, stats package]: Draws a simple heatmap</li>
<li>heatmap.2() [gplots R package]: Draws an enhanced heatmap compared to the R base function.</li>
<li>pheatmap() [pheatmap R package]: Draws pretty heatmaps and provides more control to change the appearance of heatmaps.</li>
<li>d3heatmap() [d3heatmap R package]: Draws an interactive/clickable heatmap</li>
<li>Heatmap() [ComplexHeatmap R/Bioconductor package]: Draws, annotates and arranges complex heatmaps (very useful for genomic data analysis)</li>
</ul>
<h3 id="data-preparation"><a href="#data-preparation" class="headerlink" title="data preparation"></a>data preparation</h3><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># start by standardizing the data to make variables comparable</span></span><br><span class="line">df &lt;- scale(mtcars)</span><br></pre></td></tr></tbody></table></figure>

<h3 id="R-base-heatmap-heatmap"><a href="#R-base-heatmap-heatmap" class="headerlink" title="R base heatmap: heatmap()"></a>R base heatmap: heatmap()</h3><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">heatmap(df, scale = <span class="string">"row"</span>)</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_48_0.png" alt="png"></p>
<p>It’s possible to specify a color palette using the argument col, which can be defined as follow:</p>
<ul>
<li>use custom colors</li>
</ul>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">col&lt;- colorRampPalette(<span class="built_in">c</span>(<span class="string">"red"</span>, <span class="string">"white"</span>, <span class="string">"blue"</span>))(<span class="number">256</span>)</span><br></pre></td></tr></tbody></table></figure>

<ul>
<li>Or, using RColorBrewer color palette:</li>
</ul>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">library(<span class="string">"RColorBrewer"</span>)</span><br><span class="line">col &lt;- colorRampPalette(brewer.pal(<span class="number">10</span>, <span class="string">"RdYlBu"</span>))(<span class="number">256</span>)</span><br></pre></td></tr></tbody></table></figure>

<p>Additionally, you can use the argument RowSideColors and ColSideColors to annotate rows and columns, respectively.</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Use RColorBrewer color palette names</span></span><br><span class="line">library(<span class="string">"RColorBrewer"</span>)</span><br><span class="line">col &lt;- colorRampPalette(brewer.pal(<span class="number">10</span>, <span class="string">"RdYlBu"</span>))(<span class="number">256</span>)</span><br><span class="line">heatmap(df, scale = <span class="string">"none"</span>, col =  col, </span><br><span class="line">        RowSideColors = <span class="built_in">rep</span>(<span class="built_in">c</span>(<span class="string">"blue"</span>, <span class="string">"pink"</span>), each = <span class="number">16</span>),</span><br><span class="line">        ColSideColors = <span class="built_in">c</span>(<span class="built_in">rep</span>(<span class="string">"purple"</span>, <span class="number">5</span>), <span class="built_in">rep</span>(<span class="string">"orange"</span>, <span class="number">6</span>)))</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_55_0.png" alt="png"></p>
<h3 id="Enhanced-heat-maps-heatmap-2"><a href="#Enhanced-heat-maps-heatmap-2" class="headerlink" title="Enhanced heat maps: heatmap.2()"></a>Enhanced heat maps: heatmap.2()</h3><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># install.packages("gplots")</span></span><br><span class="line">suppressPackageStartupMessages(library(<span class="string">"gplots"</span>))</span><br><span class="line">heatmap.2(df, scale = <span class="string">"none"</span>, col = bluered(<span class="number">100</span>), </span><br><span class="line">          trace = <span class="string">"none"</span>, density.info = <span class="string">"none"</span>)</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_57_0.png" alt="png"></p>
<h3 id="Pretty-heat-maps-pheatmap"><a href="#Pretty-heat-maps-pheatmap" class="headerlink" title="Pretty heat maps: pheatmap()"></a>Pretty heat maps: pheatmap()</h3><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">library(<span class="string">"pheatmap"</span>)</span><br><span class="line">pheatmap(df, cutree_rows = <span class="number">4</span>, cutree_cols = <span class="number">2</span>)</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_59_0.png" alt="png"></p>
<h3 id="Interactive-heat-maps-d3heatmap"><a href="#Interactive-heat-maps-d3heatmap" class="headerlink" title="Interactive heat maps: d3heatmap()"></a>Interactive heat maps: d3heatmap()</h3><h3 id="Complex-heatmap"><a href="#Complex-heatmap" class="headerlink" title="Complex heatmap"></a>Complex heatmap</h3><h4 id="simple-heatmap"><a href="#simple-heatmap" class="headerlink" title="simple heatmap"></a>simple heatmap</h4><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">suppressPackageStartupMessages(library(ComplexHeatmap))</span><br><span class="line">Heatmap(df, </span><br><span class="line">        name = <span class="string">"mtcars"</span>, <span class="comment">#title of legend</span></span><br><span class="line">        column_title = <span class="string">"Variables"</span>, row_title = <span class="string">"Samples"</span>,</span><br><span class="line">        row_title_gp = gpar(fontsize = <span class="number">14</span>),</span><br><span class="line">        row_names_gp = gpar(fontsize = <span class="number">10</span>), <span class="comment"># Text size for row names</span></span><br><span class="line">        column_names_gp = gpar(fontsize = <span class="number">14</span>)</span><br><span class="line">        )</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_63_0.png" alt="png"></p>
<p>To specify a custom colors, you must use the the <code>colorRamp2()</code> function [circlize package], as follow</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">suppressPackageStartupMessages(library(circlize))</span><br><span class="line">mycols &lt;- colorRamp2(breaks = <span class="built_in">c</span>(-<span class="number">2</span>, <span class="number">0</span>, <span class="number">2</span>), </span><br><span class="line">                    colors = <span class="built_in">c</span>(<span class="string">"green"</span>, <span class="string">"white"</span>, <span class="string">"red"</span>))</span><br><span class="line">Heatmap(df, name = <span class="string">"mtcars"</span>, col = mycols)</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_65_0.png" alt="png"></p>
<p>We can also customize the appearance of dendograms using the function <code>color_branches()</code> [dendextend package]</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">library(dendextend)</span><br><span class="line">row_dend &lt;- hclust(dist(df))</span><br><span class="line">col_dend &lt;- hclust(dist(t(df)))</span><br><span class="line">Heatmap(df, name=<span class="string">'mtcars'</span>,</span><br><span class="line">        row_names_gp = gpar(fontsize = <span class="number">11</span>),</span><br><span class="line">        cluster_rows = color_branches(row_dend,k=<span class="number">4</span>),</span><br><span class="line">        cluster_columns = color_branches(col_dend,k=<span class="number">2</span>)</span><br><span class="line">       )</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_67_0.png" alt="png"></p>
<h4 id="Splitting-heatmap-by-rows"><a href="#Splitting-heatmap-by-rows" class="headerlink" title="Splitting heatmap by rows"></a>Splitting heatmap by rows</h4><p>You can split the heatmap using either the k-means algorithm or a grouping variable</p>
<ul>
<li>by k-means</li>
</ul>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Divide into 2 groups</span></span><br><span class="line">set.seed(<span class="number">2</span>)</span><br><span class="line">Heatmap(df, name = <span class="string">"mtcars"</span>, k = <span class="number">2</span>)</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_71_0.png" alt="png"></p>
<ul>
<li>grouping variables</li>
</ul>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># split by a vector specifying rowgroups</span></span><br><span class="line">Heatmap(df, name = <span class="string">"mtcars"</span>, split = mtcars$cyl,</span><br><span class="line">        row_names_gp = gpar(fontsize = <span class="number">7</span>))</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_73_0.png" alt="png"></p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Split by combining multiple variables</span></span><br><span class="line">Heatmap(df, name =<span class="string">"mtcars"</span>, </span><br><span class="line">        split = data.frame(cyl = mtcars$cyl, am = mtcars$am),</span><br><span class="line">        row_names_gp = gpar(fontsize = <span class="number">7</span>))</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_74_0.png" alt="png"></p>
<h3 id="Heatmap-annotation"><a href="#Heatmap-annotation" class="headerlink" title="Heatmap annotation"></a>Heatmap annotation</h3><p>For the example below, we’ll transpose our data to have the observations in columns and the variables in rows.</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">df &lt;- t(df)</span><br></pre></td></tr></tbody></table></figure>

<h4 id="Simple-annotation"><a href="#Simple-annotation" class="headerlink" title="Simple annotation"></a>Simple annotation</h4><p>A vector, containing discrete or continuous values, is used to annotate rows or columns.<br>We’ll use the qualitative variables cyl (levels = “4”, “5” and “8”) and am (levels = “0” and “1”), and the continuous variable mpg to annotate columns.</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">table(mtcars[<span class="string">'am'</span>])</span><br></pre></td></tr></tbody></table></figure>


<pre><code> 0  1 
19 13 
</code></pre>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Define colors for each levels of qualitative variables</span></span><br><span class="line">col = <span class="built_in">list</span>(cyl = <span class="built_in">c</span>(<span class="string">"4"</span>=<span class="string">"green"</span>,<span class="string">"6"</span>=<span class="string">"gray"</span>,<span class="string">"8"</span>=<span class="string">"darkred"</span>),</span><br><span class="line">          am = <span class="built_in">c</span>(<span class="string">"0"</span>=<span class="string">"yellow"</span>,<span class="string">"1"</span>=<span class="string">"orange"</span>),</span><br><span class="line">          mpg = circlize::colorRamp2(<span class="built_in">c</span>(<span class="number">10</span>,<span class="number">34</span>),</span><br><span class="line">                                    <span class="built_in">c</span>(<span class="string">"lightblue"</span>, <span class="string">"purple"</span>)))</span><br><span class="line">ha &lt;- HeatmapAnnotation(</span><br><span class="line">cyl = mtcars$cyl, am = mtcars$am, mpg = mtcars$mpg,</span><br><span class="line">col = col)</span><br></pre></td></tr></tbody></table></figure>


<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">Heatmap(df, name=<span class="string">'mtcars'</span>,</span><br><span class="line">       top_annotation = ha)</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_82_0.png" alt="png"></p>
<h4 id="complex-annotation"><a href="#complex-annotation" class="headerlink" title="*complex annotation"></a>*complex annotation</h4><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Define some graphics to display the distribution of columns</span></span><br><span class="line">.hist = anno_histogram(df, gp = gpar(fill = <span class="string">"lightblue"</span>))</span><br><span class="line">.density = anno_density(df, type = <span class="string">"line"</span>, gp = gpar(col = <span class="string">"blue"</span>))</span><br><span class="line">ha_mix_top = HeatmapAnnotation(</span><br><span class="line">  hist = .hist, density = .density,</span><br><span class="line">  height = unit(<span class="number">3.8</span>, <span class="string">"cm"</span>)</span><br><span class="line">  )</span><br><span class="line"><span class="comment"># Define some graphics to display the distribution of rows</span></span><br><span class="line">.violin = anno_density(df, type = <span class="string">"violin"</span>, </span><br><span class="line">                       gp = gpar(fill = <span class="string">"lightblue"</span>), which = <span class="string">"row"</span>)</span><br><span class="line">.boxplot = anno_boxplot(df, which = <span class="string">"row"</span>)</span><br><span class="line">ha_mix_right = HeatmapAnnotation(violin = .violin, bxplt = .boxplot,</span><br><span class="line">                              which = <span class="string">"row"</span>, width = unit(<span class="number">4</span>, <span class="string">"cm"</span>))</span><br><span class="line"><span class="comment"># Combine annotation with heatmap</span></span><br><span class="line">Heatmap(df, name = <span class="string">"mtcars"</span>, </span><br><span class="line">        column_names_gp = gpar(fontsize = <span class="number">8</span>),</span><br><span class="line">        top_annotation = ha_mix_top) + ha_mix_right</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_84_0.png" alt="png"></p>
<h4 id="Combining-multiple-heatmaps"><a href="#Combining-multiple-heatmaps" class="headerlink" title="*Combining multiple heatmaps"></a>*Combining multiple heatmaps</h4><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Heatmap 1</span></span><br><span class="line">ht1 = Heatmap(df, name = <span class="string">"ht1"</span>, km = <span class="number">2</span>,</span><br><span class="line">              column_names_gp = gpar(fontsize = <span class="number">9</span>))</span><br><span class="line"><span class="comment"># Heatmap 2</span></span><br><span class="line">ht2 = Heatmap(df, name = <span class="string">"ht2"</span>, </span><br><span class="line">        col = circlize::colorRamp2(<span class="built_in">c</span>(-<span class="number">2</span>, <span class="number">0</span>, <span class="number">2</span>), <span class="built_in">c</span>(<span class="string">"green"</span>, <span class="string">"white"</span>, <span class="string">"red"</span>)),</span><br><span class="line">        column_names_gp = gpar(fontsize = <span class="number">9</span>))</span><br><span class="line"><span class="comment"># Combine the two heatmaps</span></span><br><span class="line">ht1 + ht2</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_86_0.png" alt="png"></p>
<h3 id="Application-to-gene-expression-matrix"><a href="#Application-to-gene-expression-matrix" class="headerlink" title="Application to gene expression matrix"></a>Application to gene expression matrix</h3><p>In gene expression data, <u>rows are genes</u> and <u>columns are samples</u>.   </p>
<p>More information about genes can be attached after the expression heatmap such as gene length and type of genes.</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">expr &lt;- readRDS(paste0(system.file(package = <span class="string">"ComplexHeatmap"</span>),</span><br><span class="line">                      <span class="string">"/extdata/gene_expression.rds"</span>))</span><br></pre></td></tr></tbody></table></figure>


<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">head(expr,<span class="number">1</span>)</span><br></pre></td></tr></tbody></table></figure>


<table class="dataframe">
<caption>A data.frame: 1 × 27</caption>
<thead>
    <tr><th></th><th scope="col">s1_cell01</th><th scope="col">s2_cell02</th><th scope="col">s3_cell03</th><th scope="col">s4_cell01</th><th scope="col">s5_cell02</th><th scope="col">s6_cell03</th><th scope="col">s7_cell01</th><th scope="col">s8_cell02</th><th scope="col">s9_cell03</th><th scope="col">s10_cell01</th><th scope="col">...</th><th scope="col">s18_cell03</th><th scope="col">s19_cell01</th><th scope="col">s20_cell02</th><th scope="col">s21_cell03</th><th scope="col">s22_cell01</th><th scope="col">s23_cell02</th><th scope="col">s24_cell03</th><th scope="col">length</th><th scope="col">type</th><th scope="col">chr</th></tr>
    <tr><th></th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;dbl&gt;</th><th scope="col">...</th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;dbl&gt;</th><th scope="col">&lt;chr&gt;</th><th scope="col">&lt;chr&gt;</th></tr>
</thead>
<tbody>
    <tr><th scope="row">gene1</th><td>11.11632</td><td>12.65986</td><td>11.3805</td><td>10.84344</td><td>12.29402</td><td>11.07829</td><td>11.08335</td><td>12.83157</td><td>11.24916</td><td>10.45512</td><td>...</td><td>11.41324</td><td>10.69669</td><td>12.83305</td><td>11.12082</td><td>10.25306</td><td>12.63911</td><td>11.20286</td><td>124882</td><td>protein_coding</td><td>chr1</td></tr>
</tbody>
</table>




<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">table(expr$chr)</span><br></pre></td></tr></tbody></table></figure>


<pre><code> chr1 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr19  chr2 chr20 
    9    10     6     7     3     8     9     3     5     5     5    14     3 
chr22  chr3  chr4  chr5  chr6  chr7  chr8  chr9  chrM  chrX 
    1     7     8     9     9    12     5     4     6     7 
</code></pre>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">mat &lt;- as.matrix(expr[,grep(<span class="string">"cell"</span>,colnames(expr))])</span><br><span class="line">type &lt;- gsub(<span class="string">'s\\d+_'</span>,<span class="string">''</span>,colnames(mat))</span><br><span class="line">ha = HeatmapAnnotation(</span><br><span class="line">  df = data.frame(type = type),</span><br><span class="line">    annotation_height = unit(<span class="number">4</span>, <span class="string">"mm"</span>)</span><br><span class="line">  )</span><br><span class="line"></span><br><span class="line">Heatmap(mat, name=<span class="string">'expression'</span> , km = <span class="number">5</span>, top_annotation = ha,</span><br><span class="line">        show_column_names = <span class="built_in">F</span>, show_row_names = <span class="built_in">F</span>) + </span><br><span class="line">Heatmap(expr$<span class="built_in">length</span>, name = <span class="string">'length'</span>, width = unit(<span class="number">5</span>, <span class="string">"mm"</span>),</span><br><span class="line">        col = circlize::colorRamp2(<span class="built_in">c</span>(<span class="number">0</span>,<span class="number">100000</span>),<span class="built_in">c</span>(<span class="string">'white'</span>,<span class="string">'orange'</span>))) +</span><br><span class="line">Heatmap(expr$type, name = <span class="string">'type'</span>, width = unit(<span class="number">5</span>, <span class="string">"mm"</span>)) +</span><br><span class="line">Heatmap(expr$chr, name = <span class="string">'chr'</span>, width = unit(<span class="number">5</span>, <span class="string">"mm"</span>),</span><br><span class="line">       col = circlize::rand_color(<span class="built_in">length</span>(unique(expr$chr))))</span><br></pre></td></tr></tbody></table></figure>


<p><img src="output_92_0.png" alt="png"></p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">set.seed(<span class="number">123</span>)</span><br><span class="line">mydata &lt;- matrix(rnorm(<span class="number">200</span>), <span class="number">20</span>, <span class="number">10</span>)</span><br><span class="line">mydata[<span class="number">1</span>:<span class="number">10</span>, seq(<span class="number">1</span>, <span class="number">10</span>, <span class="number">2</span>)] = mydata[<span class="number">1</span>:<span class="number">10</span>, seq(<span class="number">1</span>, <span class="number">10</span>, <span class="number">2</span>)] + <span class="number">3</span></span><br><span class="line">mydata[<span class="number">11</span>:<span class="number">20</span>, seq(<span class="number">2</span>, <span class="number">10</span>, <span class="number">2</span>)] = mydata[<span class="number">11</span>:<span class="number">20</span>, seq(<span class="number">2</span>, <span class="number">10</span>, <span class="number">2</span>)] + <span class="number">2</span></span><br><span class="line">mydata[<span class="number">15</span>:<span class="number">20</span>, seq(<span class="number">2</span>, <span class="number">10</span>, <span class="number">2</span>)] = mydata[<span class="number">15</span>:<span class="number">20</span>, seq(<span class="number">2</span>, <span class="number">10</span>, <span class="number">2</span>)] + <span class="number">4</span></span><br><span class="line">colnames(mydata) = paste(<span class="string">"Sple"</span>, <span class="number">1</span>:<span class="number">10</span>, sep = <span class="string">""</span>)</span><br><span class="line">rownames(mydata) = paste(<span class="string">"Gene"</span>, <span class="number">1</span>:<span class="number">20</span>, sep = <span class="string">""</span>)</span><br><span class="line">head(mydata[, <span class="number">1</span>:<span class="number">4</span>], <span class="number">4</span>)</span><br></pre></td></tr></tbody></table></figure>


<table class="dataframe">
<caption>A matrix: 4 × 4 of type dbl</caption>
<thead>
    <tr><th></th><th scope="col">Sple1</th><th scope="col">Sple2</th><th scope="col">Sple3</th><th scope="col">Sple4</th></tr>
</thead>
<tbody>
    <tr><th scope="row">Gene1</th><td>2.439524</td><td>-1.0678237</td><td>2.305293</td><td> 0.3796395</td></tr>
    <tr><th scope="row">Gene2</th><td>2.769823</td><td>-0.2179749</td><td>2.792083</td><td>-0.5023235</td></tr>
    <tr><th scope="row">Gene3</th><td>4.558708</td><td>-1.0260044</td><td>1.734604</td><td>-0.3332074</td></tr>
    <tr><th scope="row">Gene4</th><td>3.070508</td><td>-0.7288912</td><td>5.168956</td><td>-1.0185754</td></tr>
</tbody>
</table>




<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">library(<span class="string">"pheatmap"</span>)</span><br><span class="line">pheatmap(mydata, scale = <span class="string">"row"</span>)</span><br></pre></td></tr></tbody></table></figure>

<pre><code>Warning message:
"package 'pheatmap' was built under R version 4.0.3"
</code></pre>
<p><img src="output_94_1.png" alt="png"></p>

                
            </div>
            <hr/>

            

    <div class="reprint" id="reprint-statement">
        
            <div class="reprint__author">
                <span class="reprint-meta" style="font-weight: bold;">
                    <i class="fas fa-user">
                        文章作者:
                    </i>
                </span>
                <span class="reprint-info">
                    <a href="/about" rel="external nofollow noreferrer">梁绍波</a>
                </span>
            </div>
            <div class="reprint__type">
                <span class="reprint-meta" style="font-weight: bold;">
                    <i class="fas fa-link">
                        文章链接:
                    </i>
                </span>
                <span class="reprint-info">
                    <a href="https://liangshaw.github.io/2021/12/30/Hierarchical_clustering_in_R/">https://liangshaw.github.io/2021/12/30/Hierarchical_clustering_in_R/</a>
                </span>
            </div>
            <div class="reprint__notice">
                <span class="reprint-meta" style="font-weight: bold;">
                    <i class="fas fa-copyright">
                        版权声明:
                    </i>
                </span>
                <span class="reprint-info">
                    本博客所有文章除特別声明外，均采用
                    <a href="https://creativecommons.org/licenses/by/4.0/deed.zh" rel="external nofollow noreferrer" target="_blank">CC BY 4.0</a>
                    许可协议。转载请注明来源
                    <a href="/about" target="_blank">梁绍波</a>
                    !
                </span>
            </div>
        
    </div>

    <script async defer>
      document.addEventListener("copy", function (e) {
        let toastHTML = '<span>复制成功，请遵循本文的转载规则</span><button class="btn-flat toast-action" onclick="navToReprintStatement()" style="font-size: smaller">查看</a>';
        M.toast({html: toastHTML})
      });

      function navToReprintStatement() {
        $("html, body").animate({scrollTop: $("#reprint-statement").offset().top - 80}, 800);
      }
    </script>



            <div class="tag_share" style="display: block;">
                <div class="post-meta__tag-list" style="display: inline-block;">
                    
                        <div class="article-tag">
                            
                                <a href="/tags/hierarchical-clustering/">
                                    <span class="chip bg-color">hierarchical clustering</span>
                                </a>
                            
                                <a href="/tags/R/">
                                    <span class="chip bg-color">R</span>
                                </a>
                            
                        </div>
                    
                </div>
                <div class="post_share" style="zoom: 80%; width: fit-content; display: inline-block; float: right; margin: -0.15rem 0;">
                    <link rel="stylesheet" type="text/css" href="/libs/share/css/share.min.css">
<div id="article-share">

    
    <div class="social-share" data-sites="twitter,facebook,google,qq,qzone,wechat,weibo,douban,linkedin" data-wechat-qrcode-helper="<p>微信扫一扫即可分享！</p>"></div>
    <script src="/libs/share/js/social-share.min.js"></script>
    

    

</div>

                </div>
            </div>
            
        </div>
    </div>

    

    

    

    

    

    

    

    

    

    

<article id="prenext-posts" class="prev-next articles">
    <div class="row article-row">
        
        <div class="article col s12 m6" data-aos="fade-up">
            <div class="article-badge left-badge text-color">
                <i class="fas fa-chevron-left"></i>&nbsp;上一篇</div>
            <div class="card">
                <a href="/2021/12/30/Git_instructions/">
                    <div class="card-image">
                        
                        
                        <img src="/medias/featureimages/10.jpg" class="responsive-img" alt="git/github basic usage guide">
                        
                        <span class="card-title">git/github basic usage guide</span>
                    </div>
                </a>
                <div class="card-content article-content">
                    <div class="summary block-with-text">
                        
                            
                        
                    </div>
                    <div class="publish-info">
                        <span class="publish-date">
                            <i class="far fa-clock fa-fw icon-date"></i>2021-12-30
                        </span>
                        <span class="publish-author">
                            
                            <i class="fas fa-user fa-fw"></i>
                            梁绍波
                            
                        </span>
                    </div>
                </div>
                
                <div class="card-action article-tags">
                    
                    <a href="/tags/git-and-github/">
                        <span class="chip bg-color">git and github</span>
                    </a>
                    
                </div>
                
            </div>
        </div>
        
        
        <div class="article col s12 m6" data-aos="fade-up">
            <div class="article-badge right-badge text-color">
                下一篇&nbsp;<i class="fas fa-chevron-right"></i>
            </div>
            <div class="card">
                <a href="/2021/12/30/GSEA_instruction/">
                    <div class="card-image">
                        
                        
                        <img src="/medias/featureimages/24.jpg" class="responsive-img" alt="GSEA富集分析笔记">
                        
                        <span class="card-title">GSEA富集分析笔记</span>
                    </div>
                </a>
                <div class="card-content article-content">
                    <div class="summary block-with-text">
                        
                            
                        
                    </div>
                    <div class="publish-info">
                            <span class="publish-date">
                                <i class="far fa-clock fa-fw icon-date"></i>2021-12-30
                            </span>
                        <span class="publish-author">
                            
                            <i class="fas fa-user fa-fw"></i>
                            梁绍波
                            
                        </span>
                    </div>
                </div>
                
                <div class="card-action article-tags">
                    
                    <a href="/tags/GSEA/">
                        <span class="chip bg-color">GSEA</span>
                    </a>
                    
                </div>
                
            </div>
        </div>
        
    </div>
</article>

</div>



<!-- 代码块功能依赖 -->
<script type="text/javascript" src="/libs/codeBlock/codeBlockFuction.js"></script>

<!-- 代码语言 -->

<script type="text/javascript" src="/libs/codeBlock/codeLang.js"></script>


<!-- 代码块复制 -->

<script type="text/javascript" src="/libs/codeBlock/codeCopy.js"></script>


<!-- 代码块收缩 -->

<script type="text/javascript" src="/libs/codeBlock/codeShrink.js"></script>


    </div>
    <div id="toc-aside" class="expanded col l3 hide-on-med-and-down">
        <div class="toc-widget card" style="background-color: white;">
            <div class="toc-title"><i class="far fa-list-alt"></i>&nbsp;&nbsp;目录</div>
            <div id="toc-content"></div>
        </div>
    </div>
</div>

<!-- TOC 悬浮按钮. -->

<div id="floating-toc-btn" class="hide-on-med-and-down">
    <a class="btn-floating btn-large bg-color">
        <i class="fas fa-list-ul"></i>
    </a>
</div>


<script src="/libs/tocbot/tocbot.min.js"></script>
<script>
    $(function () {
        tocbot.init({
            tocSelector: '#toc-content',
            contentSelector: '#articleContent',
            headingsOffset: -($(window).height() * 0.4 - 45),
            collapseDepth: Number('0'),
            headingSelector: 'h2, h3, h4'
        });

        // modify the toc link href to support Chinese.
        let i = 0;
        let tocHeading = 'toc-heading-';
        $('#toc-content a').each(function () {
            $(this).attr('href', '#' + tocHeading + (++i));
        });

        // modify the heading title id to support Chinese.
        i = 0;
        $('#articleContent').children('h2, h3, h4').each(function () {
            $(this).attr('id', tocHeading + (++i));
        });

        // Set scroll toc fixed.
        let tocHeight = parseInt($(window).height() * 0.4 - 64);
        let $tocWidget = $('.toc-widget');
        $(window).scroll(function () {
            let scroll = $(window).scrollTop();
            /* add post toc fixed. */
            if (scroll > tocHeight) {
                $tocWidget.addClass('toc-fixed');
            } else {
                $tocWidget.removeClass('toc-fixed');
            }
        });

        
        /* 修复文章卡片 div 的宽度. */
        let fixPostCardWidth = function (srcId, targetId) {
            let srcDiv = $('#' + srcId);
            if (srcDiv.length === 0) {
                return;
            }

            let w = srcDiv.width();
            if (w >= 450) {
                w = w + 21;
            } else if (w >= 350 && w < 450) {
                w = w + 18;
            } else if (w >= 300 && w < 350) {
                w = w + 16;
            } else {
                w = w + 14;
            }
            $('#' + targetId).width(w);
        };

        // 切换TOC目录展开收缩的相关操作.
        const expandedClass = 'expanded';
        let $tocAside = $('#toc-aside');
        let $mainContent = $('#main-content');
        $('#floating-toc-btn .btn-floating').click(function () {
            if ($tocAside.hasClass(expandedClass)) {
                $tocAside.removeClass(expandedClass).hide();
                $mainContent.removeClass('l9');
            } else {
                $tocAside.addClass(expandedClass).show();
                $mainContent.addClass('l9');
            }
            fixPostCardWidth('artDetail', 'prenext-posts');
        });
        
    });
</script>

    

</main>




    <footer class="page-footer bg-color">
    
        <link rel="stylesheet" href="/libs/aplayer/APlayer.min.css">
<style>
    .aplayer .aplayer-lrc p {
        
        display: none;
        
        font-size: 12px;
        font-weight: 700;
        line-height: 16px !important;
    }

    .aplayer .aplayer-lrc p.aplayer-lrc-current {
        
        display: none;
        
        font-size: 15px;
        color: #42b983;
    }

    
    .aplayer.aplayer-fixed.aplayer-narrow .aplayer-body {
        left: -66px !important;
    }

    .aplayer.aplayer-fixed.aplayer-narrow .aplayer-body:hover {
        left: 0px !important;
    }

    
</style>
<div class="">
    
    <div class="row">
        <meting-js class="col l8 offset-l2 m10 offset-m1 s12"
                   server="netease"
                   type="playlist"
                   id="3171388577"
                   fixed='true'
                   autoplay='true'
                   theme='#42b983'
                   loop='all'
                   order='list'
                   preload='auto'
                   volume='0.5'
                   list-folded='true'
        >
        </meting-js>
    </div>
</div>

<script src="/libs/aplayer/APlayer.min.js"></script>
<script src="/libs/aplayer/Meting.min.js"></script>

    

    <div class="container row center-align"
         style="margin-bottom: 15px !important;">
        <div class="col s12 m8 l8 copy-right">
            Copyright&nbsp;&copy;
            
                <span id="year">2022-2021</span>
            
            <a href="/about" target="_blank">梁绍波</a>
            |&nbsp;Powered by&nbsp;<a href="https://hexo.io/" target="_blank">Hexo</a>
            |&nbsp;Theme&nbsp;<a href="https://github.com/blinkfox/hexo-theme-matery" target="_blank">Matery</a>
            <br>
            
            
            
                
            
            
                <span id="busuanzi_container_site_pv">
                &nbsp;|&nbsp;<i class="far fa-eye"></i>&nbsp;总访问量:&nbsp;
                    <span id="busuanzi_value_site_pv" class="white-color"></span>
            </span>
            
            
                <span id="busuanzi_container_site_uv">
                &nbsp;|&nbsp;<i class="fas fa-users"></i>&nbsp;总访问人数:&nbsp;
                    <span id="busuanzi_value_site_uv" class="white-color"></span>
            </span>
            
            <br>

            <!-- 运行天数提醒. -->
            
                <span id="sitetime"> Loading ...</span>
                <script>
                    var calcSiteTime = function () {
                        var seconds = 1000;
                        var minutes = seconds * 60;
                        var hours = minutes * 60;
                        var days = hours * 24;
                        var years = days * 365;
                        var today = new Date();
                        var startYear = "2022";
                        var startMonth = "1";
                        var startDate = "1";
                        var startHour = "0";
                        var startMinute = "0";
                        var startSecond = "0";
                        var todayYear = today.getFullYear();
                        var todayMonth = today.getMonth() + 1;
                        var todayDate = today.getDate();
                        var todayHour = today.getHours();
                        var todayMinute = today.getMinutes();
                        var todaySecond = today.getSeconds();
                        var t1 = Date.UTC(startYear, startMonth, startDate, startHour, startMinute, startSecond);
                        var t2 = Date.UTC(todayYear, todayMonth, todayDate, todayHour, todayMinute, todaySecond);
                        var diff = t2 - t1;
                        var diffYears = Math.floor(diff / years);
                        var diffDays = Math.floor((diff / days) - diffYears * 365);

                        // 区分是否有年份.
                        var language = 'zh-CN';
                        if (startYear === String(todayYear)) {
                            document.getElementById("year").innerHTML = todayYear;
                            var daysTip = 'This site has been running for ' + diffDays + ' days';
                            if (language === 'zh-CN') {
                                daysTip = '本站已运行 ' + diffDays + ' 天';
                            } else if (language === 'zh-HK') {
                                daysTip = '本站已運行 ' + diffDays + ' 天';
                            }
                            document.getElementById("sitetime").innerHTML = daysTip;
                        } else {
                            document.getElementById("year").innerHTML = startYear + " - " + todayYear;
                            var yearsAndDaysTip = 'This site has been running for ' + diffYears + ' years and '
                                + diffDays + ' days';
                            if (language === 'zh-CN') {
                                yearsAndDaysTip = '本站已运行 ' + diffYears + ' 年 ' + diffDays + ' 天';
                            } else if (language === 'zh-HK') {
                                yearsAndDaysTip = '本站已運行 ' + diffYears + ' 年 ' + diffDays + ' 天';
                            }
                            document.getElementById("sitetime").innerHTML = yearsAndDaysTip;
                        }
                    }

                    calcSiteTime();
                </script>
            
            <br>
            
        </div>
        <div class="col s12 m4 l4 social-link social-statis">
    <a href="https://github.com/LiangShaw" class="tooltipped" target="_blank" data-tooltip="访问我的GitHub" data-position="top" data-delay="50">
        <i class="fab fa-github"></i>
    </a>



    <a href="mailto:shaoboliang@whu.edu.cn" class="tooltipped" target="_blank" data-tooltip="邮件联系我" data-position="top" data-delay="50">
        <i class="fas fa-envelope-open"></i>
    </a>







    <a href="tencent://AddContact/?fromId=50&fromSubId=1&subcmd=all&uin=1323909095" class="tooltipped" target="_blank" data-tooltip="QQ联系我: 1323909095" data-position="top" data-delay="50">
        <i class="fab fa-qq"></i>
    </a>







</div>
    </div>
</footer>

<div class="progress-bar"></div>


    <!-- 搜索遮罩框 -->
<div id="searchModal" class="modal">
    <div class="modal-content">
        <div class="search-header">
            <span class="title"><i class="fas fa-search"></i>&nbsp;&nbsp;搜索</span>
            <input type="search" id="searchInput" name="s" placeholder="请输入搜索的关键字"
                   class="search-input">
        </div>
        <div id="searchResult"></div>
    </div>
</div>

<script type="text/javascript">
$(function () {
    var searchFunc = function (path, search_id, content_id) {
        'use strict';
        $.ajax({
            url: path,
            dataType: "xml",
            success: function (xmlResponse) {
                // get the contents from search data
                var datas = $("entry", xmlResponse).map(function () {
                    return {
                        title: $("title", this).text(),
                        content: $("content", this).text(),
                        url: $("url", this).text()
                    };
                }).get();
                var $input = document.getElementById(search_id);
                var $resultContent = document.getElementById(content_id);
                $input.addEventListener('input', function () {
                    var str = '<ul class=\"search-result-list\">';
                    var keywords = this.value.trim().toLowerCase().split(/[\s\-]+/);
                    $resultContent.innerHTML = "";
                    if (this.value.trim().length <= 0) {
                        return;
                    }
                    // perform local searching
                    datas.forEach(function (data) {
                        var isMatch = true;
                        var data_title = data.title.trim().toLowerCase();
                        var data_content = data.content.trim().replace(/<[^>]+>/g, "").toLowerCase();
                        var data_url = data.url;
                        data_url = data_url.indexOf('/') === 0 ? data.url : '/' + data_url;
                        var index_title = -1;
                        var index_content = -1;
                        var first_occur = -1;
                        // only match artiles with not empty titles and contents
                        if (data_title !== '' && data_content !== '') {
                            keywords.forEach(function (keyword, i) {
                                index_title = data_title.indexOf(keyword);
                                index_content = data_content.indexOf(keyword);
                                if (index_title < 0 && index_content < 0) {
                                    isMatch = false;
                                } else {
                                    if (index_content < 0) {
                                        index_content = 0;
                                    }
                                    if (i === 0) {
                                        first_occur = index_content;
                                    }
                                }
                            });
                        }
                        // show search results
                        if (isMatch) {
                            str += "<li><a href='" + data_url + "' class='search-result-title'>" + data_title + "</a>";
                            var content = data.content.trim().replace(/<[^>]+>/g, "");
                            if (first_occur >= 0) {
                                // cut out 100 characters
                                var start = first_occur - 20;
                                var end = first_occur + 80;
                                if (start < 0) {
                                    start = 0;
                                }
                                if (start === 0) {
                                    end = 100;
                                }
                                if (end > content.length) {
                                    end = content.length;
                                }
                                var match_content = content.substr(start, end);
                                // highlight all keywords
                                keywords.forEach(function (keyword) {
                                    var regS = new RegExp(keyword, "gi");
                                    match_content = match_content.replace(regS, "<em class=\"search-keyword\">" + keyword + "</em>");
                                });

                                str += "<p class=\"search-result\">" + match_content + "...</p>"
                            }
                            str += "</li>";
                        }
                    });
                    str += "</ul>";
                    $resultContent.innerHTML = str;
                });
            }
        });
    };

    searchFunc('/search.xml', 'searchInput', 'searchResult');
});
</script>

    <!-- 回到顶部按钮 -->
<div id="backTop" class="top-scroll">
    <a class="btn-floating btn-large waves-effect waves-light" href="#!">
        <i class="fas fa-arrow-up"></i>
    </a>
</div>


    <script src="/libs/materialize/materialize.min.js"></script>
    <script src="/libs/masonry/masonry.pkgd.min.js"></script>
    <script src="/libs/aos/aos.js"></script>
    <script src="/libs/scrollprogress/scrollProgress.min.js"></script>
    <script src="/libs/lightGallery/js/lightgallery-all.min.js"></script>
    <script src="/js/matery.js"></script>

    

    

    <!-- 雪花特效 -->
    

    <!-- 鼠标星星特效 -->
    

     
        <script src="https://ssl.captcha.qq.com/TCaptcha.js"></script>
        <script src="/libs/others/TencentCaptcha.js"></script>
        <button id="TencentCaptcha" data-appid="xxxxxxxxxx" data-cbfn="callback" type="button" hidden></button>
    

    <!-- Baidu Analytics -->

    <!-- Baidu Push -->

<script>
    (function () {
        var bp = document.createElement('script');
        var curProtocol = window.location.protocol.split(':')[0];
        if (curProtocol === 'https') {
            bp.src = 'https://zz.bdstatic.com/linksubmit/push.js';
        } else {
            bp.src = 'http://push.zhanzhang.baidu.com/push.js';
        }
        var s = document.getElementsByTagName("script")[0];
        s.parentNode.insertBefore(bp, s);
    })();
</script>

    
    <script src="/libs/others/clicklove.js" async="async"></script>
    
    
    <script async src="/libs/others/busuanzi.pure.mini.js"></script>
    

    

    

    <!--腾讯兔小巢-->
    
    

    

    

    
    <script src="/libs/instantpage/instantpage.js" type="module"></script>
    

</body>

</html>
