<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>Hexo</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="description" content="HIERARCHICAL CLUSTERING IN R: THE ESSENTIALS (层级聚类)link1-lessonlink2-heatmap-extend Agglomerative Hierarchical Clustering丛集聚类分裂聚类    Steps to agglomerative hierarchical clustering  Data structure and">
<meta property="og:type" content="article">
<meta property="og:title" content="Hexo">
<meta property="og:url" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/index.html">
<meta property="og:site_name" content="Hexo">
<meta property="og:description" content="HIERARCHICAL CLUSTERING IN R: THE ESSENTIALS (层级聚类)link1-lessonlink2-heatmap-extend Agglomerative Hierarchical Clustering丛集聚类分裂聚类    Steps to agglomerative hierarchical clustering  Data structure and">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/hierarchical_clustering.jpg">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/Steps_to_agglomerative.jpg">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_18_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_20_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_27_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_31_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_41_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_48_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_55_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_57_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_59_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_63_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_65_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_67_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_71_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_73_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_74_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_82_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_84_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_86_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_92_0.png">
<meta property="og:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/output_94_1.png">
<meta property="article:published_time" content="2021-12-30T11:39:13.434Z">
<meta property="article:modified_time" content="2021-12-30T11:39:13.434Z">
<meta property="article:author" content="shaoboliang">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://example.com/2021/12/30/Hierarchical_clustering_in_R/img/hierarchical_clustering.jpg">
  
    <link rel="alternate" href="/atom.xml" title="Hexo" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
<meta name="generator" content="Hexo 6.0.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">Hexo</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
          <a id="nav-rss-link" class="nav-icon" href="/atom.xml" title="RSS Feed"></a>
        
        <a id="nav-search-btn" class="nav-icon" title="Search"></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="Search"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://example.com"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main"><article id="post-Hierarchical_clustering_in_R" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/12/30/Hierarchical_clustering_in_R/" class="article-date">
  <time class="dt-published" datetime="2021-12-30T11:39:13.434Z" itemprop="datePublished">2021-12-30</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="HIERARCHICAL-CLUSTERING-IN-R-THE-ESSENTIALS-层级聚类"><a href="#HIERARCHICAL-CLUSTERING-IN-R-THE-ESSENTIALS-层级聚类" class="headerlink" title="HIERARCHICAL CLUSTERING IN R: THE ESSENTIALS (层级聚类)"></a>HIERARCHICAL CLUSTERING IN R: THE ESSENTIALS (层级聚类)</h1><p><a target="_blank" rel="noopener" href="https://www.datanovia.com/en/lessons/agglomerative-hierarchical-clustering/">link1-lesson</a><br><a target="_blank" rel="noopener" href="https://www.datanovia.com/en/blog/clustering-using-correlation-as-distance-measures-in-r/">link2-heatmap-extend</a></p>
<h2 id="Agglomerative-Hierarchical-Clustering"><a href="#Agglomerative-Hierarchical-Clustering" class="headerlink" title="Agglomerative Hierarchical Clustering"></a>Agglomerative Hierarchical Clustering</h2><p><a target="_blank" rel="noopener" href="https://www.datanovia.com/en/lessons/agglomerative-hierarchical-clustering/">丛集聚类</a><br>分裂聚类</p>
<img src='img/hierarchical_clustering.jpg' width=600 />


<h2 id="Steps-to-agglomerative-hierarchical-clustering"><a href="#Steps-to-agglomerative-hierarchical-clustering" class="headerlink" title="Steps to agglomerative hierarchical clustering"></a>Steps to agglomerative hierarchical clustering</h2><center><img src='img/Steps_to_agglomerative.jpg' width=800 /></center>

<h3 id="Data-structure-and-preparation"><a href="#Data-structure-and-preparation" class="headerlink" title="Data structure and preparation"></a>Data structure and preparation</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Load the data</span></span><br><span class="line">data(<span class="string">&quot;USArrests&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Standardize the data</span></span><br><span class="line"><span class="comment">## The R function scale() can be used for standardization</span></span><br><span class="line">df &lt;- scale(USArrests)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Show the first 6 rows</span></span><br><span class="line">head(df, nrow = <span class="number">6</span>)</span><br></pre></td></tr></table></figure>


<table class="dataframe">
<caption>A matrix: 6 × 4 of type dbl</caption>
<thead>
    <tr><th></th><th scope=col>Murder</th><th scope=col>Assault</th><th scope=col>UrbanPop</th><th scope=col>Rape</th></tr>
</thead>
<tbody>
    <tr><th scope=row>Alabama</th><td>1.24256408</td><td>0.7828393</td><td>-0.5209066</td><td>-0.003416473</td></tr>
    <tr><th scope=row>Alaska</th><td>0.50786248</td><td>1.1068225</td><td>-1.2117642</td><td> 2.484202941</td></tr>
    <tr><th scope=row>Arizona</th><td>0.07163341</td><td>1.4788032</td><td> 0.9989801</td><td> 1.042878388</td></tr>
    <tr><th scope=row>Arkansas</th><td>0.23234938</td><td>0.2308680</td><td>-1.0735927</td><td>-0.184916602</td></tr>
    <tr><th scope=row>California</th><td>0.27826823</td><td>1.2628144</td><td> 1.7589234</td><td> 2.067820292</td></tr>
    <tr><th scope=row>Colorado</th><td>0.02571456</td><td>0.3988593</td><td> 0.8608085</td><td> 1.864967207</td></tr>
</tbody>
</table>



<h3 id="Similarity-measures"><a href="#Similarity-measures" class="headerlink" title="Similarity measures"></a>Similarity measures</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Compute the dissimilarity matrix</span></span><br><span class="line"><span class="comment"># df = the standardized data</span></span><br><span class="line">res.dist &lt;- dist(df, method = <span class="string">&quot;euclidean&quot;</span>)  <span class="comment"># Euclidean and manhattan</span></span><br></pre></td></tr></table></figure>


<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># view distance</span></span><br><span class="line">as.matrix(res.dist)[<span class="number">1</span>:<span class="number">6</span>, <span class="number">1</span>:<span class="number">6</span>]</span><br></pre></td></tr></table></figure>


<table class="dataframe">
<caption>A matrix: 6 × 6 of type dbl</caption>
<thead>
    <tr><th></th><th scope=col>Alabama</th><th scope=col>Alaska</th><th scope=col>Arizona</th><th scope=col>Arkansas</th><th scope=col>California</th><th scope=col>Colorado</th></tr>
</thead>
<tbody>
    <tr><th scope=row>Alabama</th><td>0.000000</td><td>2.703754</td><td>2.293520</td><td>1.289810</td><td>3.263110</td><td>2.651067</td></tr>
    <tr><th scope=row>Alaska</th><td>2.703754</td><td>0.000000</td><td>2.700643</td><td>2.826039</td><td>3.012541</td><td>2.326519</td></tr>
    <tr><th scope=row>Arizona</th><td>2.293520</td><td>2.700643</td><td>0.000000</td><td>2.717758</td><td>1.310484</td><td>1.365031</td></tr>
    <tr><th scope=row>Arkansas</th><td>1.289810</td><td>2.826039</td><td>2.717758</td><td>0.000000</td><td>3.763641</td><td>2.831051</td></tr>
    <tr><th scope=row>California</th><td>3.263110</td><td>3.012541</td><td>1.310484</td><td>3.763641</td><td>0.000000</td><td>1.287619</td></tr>
    <tr><th scope=row>Colorado</th><td>2.651067</td><td>2.326519</td><td>1.365031</td><td>2.831051</td><td>1.287619</td><td>0.000000</td></tr>
</tbody>
</table>



<h3 id="Linkage"><a href="#Linkage" class="headerlink" title="Linkage"></a>Linkage</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">res.hc &lt;- hclust(d = res.dist, method = <span class="string">&quot;ward.D2&quot;</span>)</span><br></pre></td></tr></table></figure>

<h3 id="Verify-the-cluster-tree"><a href="#Verify-the-cluster-tree" class="headerlink" title="Verify the cluster tree"></a>Verify the cluster tree</h3><p>One way to measure how well the cluster tree generated by the hclust() function reflects your data is to compute the correlation between the cophenetic distances and the original distance data generated by the dist() function.<br>If the clustering is valid, the linking of objects in the cluster tree should have a strong correlation with the distances between objects in the original distance matrix.<br>评估聚类效果的方法之一，就是评估 <code>hclust()</code> 算出来的距离 和 用<code>dist()</code>算的原始数据的距离，二者之间的相关性好不好。</p>
<p>The closer the value of the correlation coefficient is to 1, the more accurately the clustering solution reflects your data. Values above 0.75 are felt to be good.<br>The “average” linkage method appears to produce high values of this statistic.<br>相关性越接近 <strong>1</strong>，说明聚类的效果越能反映真实情况.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Compute cophentic distance</span></span><br><span class="line">res.coph &lt;- cophenetic(res.hc)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Correlation between cophenetic distance and</span></span><br><span class="line"><span class="comment"># the original distance</span></span><br><span class="line">cor(res.dist, res.coph)</span><br></pre></td></tr></table></figure>


<p>0.697526563237039</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Execute the hclust() function again using the average linkage method</span></span><br><span class="line">res.hc2 &lt;- hclust(res.dist, method = <span class="string">&quot;average&quot;</span>)</span><br><span class="line"></span><br><span class="line">cor(res.dist, cophenetic(res.hc2))</span><br></pre></td></tr></table></figure>


<p>0.718038237932047</p>
<h3 id="Dendrogram-树状图"><a href="#Dendrogram-树状图" class="headerlink" title="Dendrogram(树状图)"></a>Dendrogram(树状图)</h3><p>Dendrogram can be produced in R using the base function <code>plot(res.hc)</code>, where res.hc is the output of hclust()</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plot(res.hc, cex = <span class="number">0.8</span>)</span><br></pre></td></tr></table></figure>


<p><img src="img/output_18_0.png" alt="png"></p>
<p> <code>fviz_dend()</code> ( in <code>factoextra</code> R package ) to produce a beautiful dendrogram.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># cex: label size</span></span><br><span class="line">suppressPackageStartupMessages(library(<span class="string">&quot;factoextra&quot;</span>))</span><br><span class="line">fviz_dend(res.hc, title=<span class="string">&#x27;&#x27;</span>, cex = <span class="number">0.7</span>)</span><br></pre></td></tr></table></figure>


<p><img src="img/output_20_0.png" alt="png"></p>
<h3 id="Cut-the-dendrogram-into-different-groups"><a href="#Cut-the-dendrogram-into-different-groups" class="headerlink" title="Cut the dendrogram into different groups"></a>Cut the dendrogram into different groups</h3><p>You can cut the hierarchical tree at a given height in order to partition your data into clusters.<br>The R base function cutree() can be used to cut a tree, generated by the hclust() function, into several groups either by specifying the desired number of groups or the cut height.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Cut tree into 4 groups</span></span><br><span class="line">grp &lt;- cutree(res.hc, k = <span class="number">4</span>)</span><br><span class="line">head(grp, n = <span class="number">4</span>)</span><br></pre></td></tr></table></figure>


<style>
.dl-inline {width: auto; margin:0; padding: 0}
.dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}
.dl-inline>dt::after {content: ":\0020"; padding-right: .5ex}
.dl-inline>dt:not(:first-of-type) {padding-left: .5ex}
</style><dl class=dl-inline><dt>Alabama</dt><dd>1</dd><dt>Alaska</dt><dd>2</dd><dt>Arizona</dt><dd>2</dd><dt>Arkansas</dt><dd>3</dd></dl>




<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Number of members in each cluster</span></span><br><span class="line">table(grp)</span><br></pre></td></tr></table></figure>


<pre><code>grp
 1  2  3  4 
 7 12 19 12 
</code></pre>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Get the names for the members of cluster 1</span></span><br><span class="line">rownames(df)[grp == <span class="number">1</span>]</span><br></pre></td></tr></table></figure>


<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'Alabama'</li><li>'Georgia'</li><li>'Louisiana'</li><li>'Mississippi'</li><li>'North Carolina'</li><li>'South Carolina'</li><li>'Tennessee'</li></ol>



<p><a target="_blank" rel="noopener" href="http://agroninfotech.blogspot.com/2020/06/visualizing-clusters-in-r-hierarchical.html"><strong>fviz_dend advanced handbook</strong></a></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">fviz_dend(res.hc, title=<span class="string">&#x27;&#x27;</span>, cex=<span class="number">0.7</span>, k=<span class="number">4</span>, type=<span class="string">&#x27;rectangle&#x27;</span>, rect=<span class="built_in">T</span> ,rect_border = <span class="string">&quot;jco&quot;</span>, rect_fill = <span class="literal">TRUE</span>, horiz = <span class="literal">TRUE</span>) <span class="comment"># type=&#x27;scatter&#x27;</span></span><br></pre></td></tr></table></figure>


<p><img src="img/output_27_0.png" alt="png"></p>
<h2 id="Cluster-R-package"><a href="#Cluster-R-package" class="headerlink" title="Cluster R package"></a>Cluster R package</h2><p>The R package <code>cluster</code> makes it easy to perform cluster analysis in R. It provides the function agnes() and diana() for computing agglomerative and divisive clustering, respectively. These functions perform all the necessary steps for you. You don’t need to execute the scale(), dist() and hclust() function separately.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">library(<span class="string">&quot;cluster&quot;</span>)</span><br><span class="line"><span class="comment"># Agglomerative Nesting (Hierarchical Clustering)</span></span><br><span class="line">res.agnes &lt;- agnes(x = USArrests, <span class="comment"># data matrix</span></span><br><span class="line">                   stand = <span class="literal">TRUE</span>, <span class="comment"># Standardize the data</span></span><br><span class="line">                   metric = <span class="string">&quot;euclidean&quot;</span>, <span class="comment"># metric for distance matrix</span></span><br><span class="line">                   method = <span class="string">&quot;ward&quot;</span> <span class="comment"># Linkage method</span></span><br><span class="line">                   )</span><br><span class="line"></span><br><span class="line"><span class="comment"># DIvisive ANAlysis Clustering</span></span><br><span class="line">res.diana &lt;- diana(x = USArrests, <span class="comment"># data matrix</span></span><br><span class="line">                   stand = <span class="literal">TRUE</span>, <span class="comment"># standardize the data</span></span><br><span class="line">                   metric = <span class="string">&quot;euclidean&quot;</span> <span class="comment"># metric for distance matrix</span></span><br><span class="line">                   )</span><br></pre></td></tr></table></figure>

<pre><code>Warning message:
&quot;package &#39;cluster&#39; was built under R version 4.0.5&quot;
</code></pre>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">fviz_dend(res.diana, cex = <span class="number">0.7</span>, k = <span class="number">4</span>, horiz = <span class="built_in">T</span>, type = <span class="string">&#x27;circular&#x27;</span>)</span><br></pre></td></tr></table></figure>


<p><img src="img/output_31_0.png" alt="png"></p>
<h2 id="Application-of-hierarchical-clustering-to-gene-expression-data-analysis"><a href="#Application-of-hierarchical-clustering-to-gene-expression-data-analysis" class="headerlink" title="Application of hierarchical clustering to gene expression data analysis"></a>Application of hierarchical clustering to gene expression data analysis</h2><p>In gene expression data analysis, clustering is generaly used as one of the first step to explore the data.<br>We are interested in whether there are groups of genes or groups of samples that have similar gene expression patterns.  </p>
<p>Several <a target="_blank" rel="noopener" href="https://www.datanovia.com/en/lessons/clustering-distance-measures/">clustering distance measures</a> have been described for assessing the similarity or the dissimilarity between items, in order to decide which items have to be grouped together or not.<br>These measures can be used to cluster genes or samples that are similar.</p>
<p>For most common clustering softwares, the default distance measure is the Euclidean distance.<br>The most popular methods for gene expression data are to use log2(expression + 0.25), correlation distance and complete linkage clustering agglomerative-clustering.  </p>
<p>Note that, when the data are scaled, the Euclidean distance of the z-scores is the same as correlation distance.</p>
<p>Pearson’s correlation is quite sensitive to outliers. When clustering genes, it is important to be aware of the possible impact of outliers.<br>An alternative option is to use Spearman’s correlation instead of Pearson’s correlation.</p>
<h2 id="Comparing-Cluster-Dendrograms-in-R"><a href="#Comparing-Cluster-Dendrograms-in-R" class="headerlink" title="Comparing Cluster Dendrograms in R"></a>Comparing Cluster Dendrograms in R</h2><h3 id="Data-preparation"><a href="#Data-preparation" class="headerlink" title="Data preparation"></a>Data preparation</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">df &lt;- scale(USArrests)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Subset containing 10 rows</span></span><br><span class="line">set.seed(<span class="number">123</span>)</span><br><span class="line">ss &lt;- sample(<span class="number">1</span>:<span class="number">50</span>, <span class="number">10</span>)</span><br><span class="line">df &lt;- df[ss,]</span><br></pre></td></tr></table></figure>

<h3 id="Dendrograms-comparison"><a href="#Dendrograms-comparison" class="headerlink" title="Dendrograms comparison"></a>Dendrograms comparison</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">suppressPackageStartupMessages(library(dendextend))</span><br><span class="line"></span><br><span class="line"><span class="comment"># Compute distance matrix</span></span><br><span class="line">res.dist &lt;- dist(df, method = <span class="string">&quot;euclidean&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Compute 2 hierarchical clusterings</span></span><br><span class="line">hc1 &lt;- hclust(res.dist, method = <span class="string">&quot;average&quot;</span>)</span><br><span class="line">hc2 &lt;- hclust(res.dist, method = <span class="string">&quot;ward.D2&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Create two dendrograms</span></span><br><span class="line">dend1 &lt;- as.dendrogram (hc1)</span><br><span class="line">dend2 &lt;- as.dendrogram (hc2)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Create a list to hold dendrograms</span></span><br><span class="line">dend_list &lt;- dendlist(dend1, dend2)</span><br></pre></td></tr></table></figure>

<ol>
<li>Visual comparison of two dendrograms</li>
</ol>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Align and plot two dendrograms side by side</span></span><br><span class="line">dendlist(dend1, dend2) %&gt;%</span><br><span class="line">  untangle(method = <span class="string">&quot;step1side&quot;</span>) %&gt;% <span class="comment"># Find the best alignment layout</span></span><br><span class="line">  tanglegram()                       <span class="comment"># Draw the two dendrograms</span></span><br></pre></td></tr></table></figure>


<p><img src="img/output_41_0.png" alt="png"></p>
<h2 id="Heatmap-in-R-Static-and-Interactive-Visualization"><a href="#Heatmap-in-R-Static-and-Interactive-Visualization" class="headerlink" title="Heatmap in R: Static and Interactive Visualization"></a>Heatmap in R: Static and Interactive Visualization</h2><h3 id="R-Packages-functions-for-drawing-heatmaps"><a href="#R-Packages-functions-for-drawing-heatmaps" class="headerlink" title="R Packages/functions for drawing heatmaps"></a>R Packages/functions for drawing heatmaps</h3><ul>
<li>heatmap() [R base function, stats package]: Draws a simple heatmap</li>
<li>heatmap.2() [gplots R package]: Draws an enhanced heatmap compared to the R base function.</li>
<li>pheatmap() [pheatmap R package]: Draws pretty heatmaps and provides more control to change the appearance of heatmaps.</li>
<li>d3heatmap() [d3heatmap R package]: Draws an interactive/clickable heatmap</li>
<li>Heatmap() [ComplexHeatmap R/Bioconductor package]: Draws, annotates and arranges complex heatmaps (very useful for genomic data analysis)</li>
</ul>
<h3 id="data-preparation"><a href="#data-preparation" class="headerlink" title="data preparation"></a>data preparation</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># start by standardizing the data to make variables comparable</span></span><br><span class="line">df &lt;- scale(mtcars)</span><br></pre></td></tr></table></figure>

<h3 id="R-base-heatmap-heatmap"><a href="#R-base-heatmap-heatmap" class="headerlink" title="R base heatmap: heatmap()"></a>R base heatmap: heatmap()</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">heatmap(df, scale = <span class="string">&quot;row&quot;</span>)</span><br></pre></td></tr></table></figure>


<p><img src="img/output_48_0.png" alt="png"></p>
<p>It’s possible to specify a color palette using the argument col, which can be defined as follow:</p>
<ul>
<li>use custom colors</li>
</ul>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">col&lt;- colorRampPalette(<span class="built_in">c</span>(<span class="string">&quot;red&quot;</span>, <span class="string">&quot;white&quot;</span>, <span class="string">&quot;blue&quot;</span>))(<span class="number">256</span>)</span><br></pre></td></tr></table></figure>

<ul>
<li>Or, using RColorBrewer color palette:</li>
</ul>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">library(<span class="string">&quot;RColorBrewer&quot;</span>)</span><br><span class="line">col &lt;- colorRampPalette(brewer.pal(<span class="number">10</span>, <span class="string">&quot;RdYlBu&quot;</span>))(<span class="number">256</span>)</span><br></pre></td></tr></table></figure>

<p>Additionally, you can use the argument RowSideColors and ColSideColors to annotate rows and columns, respectively.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Use RColorBrewer color palette names</span></span><br><span class="line">library(<span class="string">&quot;RColorBrewer&quot;</span>)</span><br><span class="line">col &lt;- colorRampPalette(brewer.pal(<span class="number">10</span>, <span class="string">&quot;RdYlBu&quot;</span>))(<span class="number">256</span>)</span><br><span class="line">heatmap(df, scale = <span class="string">&quot;none&quot;</span>, col =  col, </span><br><span class="line">        RowSideColors = <span class="built_in">rep</span>(<span class="built_in">c</span>(<span class="string">&quot;blue&quot;</span>, <span class="string">&quot;pink&quot;</span>), each = <span class="number">16</span>),</span><br><span class="line">        ColSideColors = <span class="built_in">c</span>(<span class="built_in">rep</span>(<span class="string">&quot;purple&quot;</span>, <span class="number">5</span>), <span class="built_in">rep</span>(<span class="string">&quot;orange&quot;</span>, <span class="number">6</span>)))</span><br></pre></td></tr></table></figure>


<p><img src="img/output_55_0.png" alt="png"></p>
<h3 id="Enhanced-heat-maps-heatmap-2"><a href="#Enhanced-heat-maps-heatmap-2" class="headerlink" title="Enhanced heat maps: heatmap.2()"></a>Enhanced heat maps: heatmap.2()</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># install.packages(&quot;gplots&quot;)</span></span><br><span class="line">suppressPackageStartupMessages(library(<span class="string">&quot;gplots&quot;</span>))</span><br><span class="line">heatmap.2(df, scale = <span class="string">&quot;none&quot;</span>, col = bluered(<span class="number">100</span>), </span><br><span class="line">          trace = <span class="string">&quot;none&quot;</span>, density.info = <span class="string">&quot;none&quot;</span>)</span><br></pre></td></tr></table></figure>


<p><img src="img/output_57_0.png" alt="png"></p>
<h3 id="Pretty-heat-maps-pheatmap"><a href="#Pretty-heat-maps-pheatmap" class="headerlink" title="Pretty heat maps: pheatmap()"></a>Pretty heat maps: pheatmap()</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">library(<span class="string">&quot;pheatmap&quot;</span>)</span><br><span class="line">pheatmap(df, cutree_rows = <span class="number">4</span>, cutree_cols = <span class="number">2</span>)</span><br></pre></td></tr></table></figure>


<p><img src="img/output_59_0.png" alt="png"></p>
<h3 id="Interactive-heat-maps-d3heatmap"><a href="#Interactive-heat-maps-d3heatmap" class="headerlink" title="Interactive heat maps: d3heatmap()"></a>Interactive heat maps: d3heatmap()</h3><h3 id="Complex-heatmap"><a href="#Complex-heatmap" class="headerlink" title="Complex heatmap"></a>Complex heatmap</h3><h4 id="simple-heatmap"><a href="#simple-heatmap" class="headerlink" title="simple heatmap"></a>simple heatmap</h4><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">suppressPackageStartupMessages(library(ComplexHeatmap))</span><br><span class="line">Heatmap(df, </span><br><span class="line">        name = <span class="string">&quot;mtcars&quot;</span>, <span class="comment">#title of legend</span></span><br><span class="line">        column_title = <span class="string">&quot;Variables&quot;</span>, row_title = <span class="string">&quot;Samples&quot;</span>,</span><br><span class="line">        row_title_gp = gpar(fontsize = <span class="number">14</span>),</span><br><span class="line">        row_names_gp = gpar(fontsize = <span class="number">10</span>), <span class="comment"># Text size for row names</span></span><br><span class="line">        column_names_gp = gpar(fontsize = <span class="number">14</span>)</span><br><span class="line">        )</span><br></pre></td></tr></table></figure>


<p><img src="img/output_63_0.png" alt="png"></p>
<p>To specify a custom colors, you must use the the <code>colorRamp2()</code> function [circlize package], as follow</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">suppressPackageStartupMessages(library(circlize))</span><br><span class="line">mycols &lt;- colorRamp2(breaks = <span class="built_in">c</span>(-<span class="number">2</span>, <span class="number">0</span>, <span class="number">2</span>), </span><br><span class="line">                    colors = <span class="built_in">c</span>(<span class="string">&quot;green&quot;</span>, <span class="string">&quot;white&quot;</span>, <span class="string">&quot;red&quot;</span>))</span><br><span class="line">Heatmap(df, name = <span class="string">&quot;mtcars&quot;</span>, col = mycols)</span><br></pre></td></tr></table></figure>


<p><img src="img/output_65_0.png" alt="png"></p>
<p>We can also customize the appearance of dendograms using the function <code>color_branches()</code> [dendextend package]</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">library(dendextend)</span><br><span class="line">row_dend &lt;- hclust(dist(df))</span><br><span class="line">col_dend &lt;- hclust(dist(t(df)))</span><br><span class="line">Heatmap(df, name=<span class="string">&#x27;mtcars&#x27;</span>,</span><br><span class="line">        row_names_gp = gpar(fontsize = <span class="number">11</span>),</span><br><span class="line">        cluster_rows = color_branches(row_dend,k=<span class="number">4</span>),</span><br><span class="line">        cluster_columns = color_branches(col_dend,k=<span class="number">2</span>)</span><br><span class="line">       )</span><br></pre></td></tr></table></figure>


<p><img src="img/output_67_0.png" alt="png"></p>
<h4 id="Splitting-heatmap-by-rows"><a href="#Splitting-heatmap-by-rows" class="headerlink" title="Splitting heatmap by rows"></a>Splitting heatmap by rows</h4><p>You can split the heatmap using either the k-means algorithm or a grouping variable</p>
<ul>
<li>by k-means</li>
</ul>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Divide into 2 groups</span></span><br><span class="line">set.seed(<span class="number">2</span>)</span><br><span class="line">Heatmap(df, name = <span class="string">&quot;mtcars&quot;</span>, k = <span class="number">2</span>)</span><br></pre></td></tr></table></figure>


<p><img src="img/output_71_0.png" alt="png"></p>
<ul>
<li>grouping variables</li>
</ul>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># split by a vector specifying rowgroups</span></span><br><span class="line">Heatmap(df, name = <span class="string">&quot;mtcars&quot;</span>, split = mtcars$cyl,</span><br><span class="line">        row_names_gp = gpar(fontsize = <span class="number">7</span>))</span><br></pre></td></tr></table></figure>


<p><img src="img/output_73_0.png" alt="png"></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Split by combining multiple variables</span></span><br><span class="line">Heatmap(df, name =<span class="string">&quot;mtcars&quot;</span>, </span><br><span class="line">        split = data.frame(cyl = mtcars$cyl, am = mtcars$am),</span><br><span class="line">        row_names_gp = gpar(fontsize = <span class="number">7</span>))</span><br></pre></td></tr></table></figure>


<p><img src="img/output_74_0.png" alt="png"></p>
<h3 id="Heatmap-annotation"><a href="#Heatmap-annotation" class="headerlink" title="Heatmap annotation"></a>Heatmap annotation</h3><p>For the example below, we’ll transpose our data to have the observations in columns and the variables in rows.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">df &lt;- t(df)</span><br></pre></td></tr></table></figure>

<h4 id="Simple-annotation"><a href="#Simple-annotation" class="headerlink" title="Simple annotation"></a>Simple annotation</h4><p>A vector, containing discrete or continuous values, is used to annotate rows or columns.<br>We’ll use the qualitative variables cyl (levels = “4”, “5” and “8”) and am (levels = “0” and “1”), and the continuous variable mpg to annotate columns.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">table(mtcars[<span class="string">&#x27;am&#x27;</span>])</span><br></pre></td></tr></table></figure>


<pre><code> 0  1 
19 13 
</code></pre>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Define colors for each levels of qualitative variables</span></span><br><span class="line">col = <span class="built_in">list</span>(cyl = <span class="built_in">c</span>(<span class="string">&quot;4&quot;</span>=<span class="string">&quot;green&quot;</span>,<span class="string">&quot;6&quot;</span>=<span class="string">&quot;gray&quot;</span>,<span class="string">&quot;8&quot;</span>=<span class="string">&quot;darkred&quot;</span>),</span><br><span class="line">          am = <span class="built_in">c</span>(<span class="string">&quot;0&quot;</span>=<span class="string">&quot;yellow&quot;</span>,<span class="string">&quot;1&quot;</span>=<span class="string">&quot;orange&quot;</span>),</span><br><span class="line">          mpg = circlize::colorRamp2(<span class="built_in">c</span>(<span class="number">10</span>,<span class="number">34</span>),</span><br><span class="line">                                    <span class="built_in">c</span>(<span class="string">&quot;lightblue&quot;</span>, <span class="string">&quot;purple&quot;</span>)))</span><br><span class="line">ha &lt;- HeatmapAnnotation(</span><br><span class="line">cyl = mtcars$cyl, am = mtcars$am, mpg = mtcars$mpg,</span><br><span class="line">col = col)</span><br></pre></td></tr></table></figure>


<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">Heatmap(df, name=<span class="string">&#x27;mtcars&#x27;</span>,</span><br><span class="line">       top_annotation = ha)</span><br></pre></td></tr></table></figure>


<p><img src="img/output_82_0.png" alt="png"></p>
<h4 id="complex-annotation"><a href="#complex-annotation" class="headerlink" title="*complex annotation"></a>*complex annotation</h4><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Define some graphics to display the distribution of columns</span></span><br><span class="line">.hist = anno_histogram(df, gp = gpar(fill = <span class="string">&quot;lightblue&quot;</span>))</span><br><span class="line">.density = anno_density(df, type = <span class="string">&quot;line&quot;</span>, gp = gpar(col = <span class="string">&quot;blue&quot;</span>))</span><br><span class="line">ha_mix_top = HeatmapAnnotation(</span><br><span class="line">  hist = .hist, density = .density,</span><br><span class="line">  height = unit(<span class="number">3.8</span>, <span class="string">&quot;cm&quot;</span>)</span><br><span class="line">  )</span><br><span class="line"><span class="comment"># Define some graphics to display the distribution of rows</span></span><br><span class="line">.violin = anno_density(df, type = <span class="string">&quot;violin&quot;</span>, </span><br><span class="line">                       gp = gpar(fill = <span class="string">&quot;lightblue&quot;</span>), which = <span class="string">&quot;row&quot;</span>)</span><br><span class="line">.boxplot = anno_boxplot(df, which = <span class="string">&quot;row&quot;</span>)</span><br><span class="line">ha_mix_right = HeatmapAnnotation(violin = .violin, bxplt = .boxplot,</span><br><span class="line">                              which = <span class="string">&quot;row&quot;</span>, width = unit(<span class="number">4</span>, <span class="string">&quot;cm&quot;</span>))</span><br><span class="line"><span class="comment"># Combine annotation with heatmap</span></span><br><span class="line">Heatmap(df, name = <span class="string">&quot;mtcars&quot;</span>, </span><br><span class="line">        column_names_gp = gpar(fontsize = <span class="number">8</span>),</span><br><span class="line">        top_annotation = ha_mix_top) + ha_mix_right</span><br></pre></td></tr></table></figure>


<p><img src="img/output_84_0.png" alt="png"></p>
<h4 id="Combining-multiple-heatmaps"><a href="#Combining-multiple-heatmaps" class="headerlink" title="*Combining multiple heatmaps"></a>*Combining multiple heatmaps</h4><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Heatmap 1</span></span><br><span class="line">ht1 = Heatmap(df, name = <span class="string">&quot;ht1&quot;</span>, km = <span class="number">2</span>,</span><br><span class="line">              column_names_gp = gpar(fontsize = <span class="number">9</span>))</span><br><span class="line"><span class="comment"># Heatmap 2</span></span><br><span class="line">ht2 = Heatmap(df, name = <span class="string">&quot;ht2&quot;</span>, </span><br><span class="line">        col = circlize::colorRamp2(<span class="built_in">c</span>(-<span class="number">2</span>, <span class="number">0</span>, <span class="number">2</span>), <span class="built_in">c</span>(<span class="string">&quot;green&quot;</span>, <span class="string">&quot;white&quot;</span>, <span class="string">&quot;red&quot;</span>)),</span><br><span class="line">        column_names_gp = gpar(fontsize = <span class="number">9</span>))</span><br><span class="line"><span class="comment"># Combine the two heatmaps</span></span><br><span class="line">ht1 + ht2</span><br></pre></td></tr></table></figure>


<p><img src="img/output_86_0.png" alt="png"></p>
<h3 id="Application-to-gene-expression-matrix"><a href="#Application-to-gene-expression-matrix" class="headerlink" title="Application to gene expression matrix"></a>Application to gene expression matrix</h3><p>In gene expression data, <u>rows are genes</u> and <u>columns are samples</u>.   </p>
<p>More information about genes can be attached after the expression heatmap such as gene length and type of genes.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">expr &lt;- readRDS(paste0(system.file(package = <span class="string">&quot;ComplexHeatmap&quot;</span>),</span><br><span class="line">                      <span class="string">&quot;/extdata/gene_expression.rds&quot;</span>))</span><br></pre></td></tr></table></figure>


<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">head(expr,<span class="number">1</span>)</span><br></pre></td></tr></table></figure>


<table class="dataframe">
<caption>A data.frame: 1 × 27</caption>
<thead>
    <tr><th></th><th scope=col>s1_cell01</th><th scope=col>s2_cell02</th><th scope=col>s3_cell03</th><th scope=col>s4_cell01</th><th scope=col>s5_cell02</th><th scope=col>s6_cell03</th><th scope=col>s7_cell01</th><th scope=col>s8_cell02</th><th scope=col>s9_cell03</th><th scope=col>s10_cell01</th><th scope=col>...</th><th scope=col>s18_cell03</th><th scope=col>s19_cell01</th><th scope=col>s20_cell02</th><th scope=col>s21_cell03</th><th scope=col>s22_cell01</th><th scope=col>s23_cell02</th><th scope=col>s24_cell03</th><th scope=col>length</th><th scope=col>type</th><th scope=col>chr</th></tr>
    <tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>...</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
    <tr><th scope=row>gene1</th><td>11.11632</td><td>12.65986</td><td>11.3805</td><td>10.84344</td><td>12.29402</td><td>11.07829</td><td>11.08335</td><td>12.83157</td><td>11.24916</td><td>10.45512</td><td>...</td><td>11.41324</td><td>10.69669</td><td>12.83305</td><td>11.12082</td><td>10.25306</td><td>12.63911</td><td>11.20286</td><td>124882</td><td>protein_coding</td><td>chr1</td></tr>
</tbody>
</table>




<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">table(expr$chr)</span><br></pre></td></tr></table></figure>


<pre><code> chr1 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr19  chr2 chr20 
    9    10     6     7     3     8     9     3     5     5     5    14     3 
chr22  chr3  chr4  chr5  chr6  chr7  chr8  chr9  chrM  chrX 
    1     7     8     9     9    12     5     4     6     7 
</code></pre>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">mat &lt;- as.matrix(expr[,grep(<span class="string">&quot;cell&quot;</span>,colnames(expr))])</span><br><span class="line">type &lt;- gsub(<span class="string">&#x27;s\\d+_&#x27;</span>,<span class="string">&#x27;&#x27;</span>,colnames(mat))</span><br><span class="line">ha = HeatmapAnnotation(</span><br><span class="line">  df = data.frame(type = type),</span><br><span class="line">    annotation_height = unit(<span class="number">4</span>, <span class="string">&quot;mm&quot;</span>)</span><br><span class="line">  )</span><br><span class="line"></span><br><span class="line">Heatmap(mat, name=<span class="string">&#x27;expression&#x27;</span> , km = <span class="number">5</span>, top_annotation = ha,</span><br><span class="line">        show_column_names = <span class="built_in">F</span>, show_row_names = <span class="built_in">F</span>) + </span><br><span class="line">Heatmap(expr$<span class="built_in">length</span>, name = <span class="string">&#x27;length&#x27;</span>, width = unit(<span class="number">5</span>, <span class="string">&quot;mm&quot;</span>),</span><br><span class="line">        col = circlize::colorRamp2(<span class="built_in">c</span>(<span class="number">0</span>,<span class="number">100000</span>),<span class="built_in">c</span>(<span class="string">&#x27;white&#x27;</span>,<span class="string">&#x27;orange&#x27;</span>))) +</span><br><span class="line">Heatmap(expr$type, name = <span class="string">&#x27;type&#x27;</span>, width = unit(<span class="number">5</span>, <span class="string">&quot;mm&quot;</span>)) +</span><br><span class="line">Heatmap(expr$chr, name = <span class="string">&#x27;chr&#x27;</span>, width = unit(<span class="number">5</span>, <span class="string">&quot;mm&quot;</span>),</span><br><span class="line">       col = circlize::rand_color(<span class="built_in">length</span>(unique(expr$chr))))</span><br></pre></td></tr></table></figure>


<p><img src="img/output_92_0.png" alt="png"></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">set.seed(<span class="number">123</span>)</span><br><span class="line">mydata &lt;- matrix(rnorm(<span class="number">200</span>), <span class="number">20</span>, <span class="number">10</span>)</span><br><span class="line">mydata[<span class="number">1</span>:<span class="number">10</span>, seq(<span class="number">1</span>, <span class="number">10</span>, <span class="number">2</span>)] = mydata[<span class="number">1</span>:<span class="number">10</span>, seq(<span class="number">1</span>, <span class="number">10</span>, <span class="number">2</span>)] + <span class="number">3</span></span><br><span class="line">mydata[<span class="number">11</span>:<span class="number">20</span>, seq(<span class="number">2</span>, <span class="number">10</span>, <span class="number">2</span>)] = mydata[<span class="number">11</span>:<span class="number">20</span>, seq(<span class="number">2</span>, <span class="number">10</span>, <span class="number">2</span>)] + <span class="number">2</span></span><br><span class="line">mydata[<span class="number">15</span>:<span class="number">20</span>, seq(<span class="number">2</span>, <span class="number">10</span>, <span class="number">2</span>)] = mydata[<span class="number">15</span>:<span class="number">20</span>, seq(<span class="number">2</span>, <span class="number">10</span>, <span class="number">2</span>)] + <span class="number">4</span></span><br><span class="line">colnames(mydata) = paste(<span class="string">&quot;Sple&quot;</span>, <span class="number">1</span>:<span class="number">10</span>, sep = <span class="string">&quot;&quot;</span>)</span><br><span class="line">rownames(mydata) = paste(<span class="string">&quot;Gene&quot;</span>, <span class="number">1</span>:<span class="number">20</span>, sep = <span class="string">&quot;&quot;</span>)</span><br><span class="line">head(mydata[, <span class="number">1</span>:<span class="number">4</span>], <span class="number">4</span>)</span><br></pre></td></tr></table></figure>


<table class="dataframe">
<caption>A matrix: 4 × 4 of type dbl</caption>
<thead>
    <tr><th></th><th scope=col>Sple1</th><th scope=col>Sple2</th><th scope=col>Sple3</th><th scope=col>Sple4</th></tr>
</thead>
<tbody>
    <tr><th scope=row>Gene1</th><td>2.439524</td><td>-1.0678237</td><td>2.305293</td><td> 0.3796395</td></tr>
    <tr><th scope=row>Gene2</th><td>2.769823</td><td>-0.2179749</td><td>2.792083</td><td>-0.5023235</td></tr>
    <tr><th scope=row>Gene3</th><td>4.558708</td><td>-1.0260044</td><td>1.734604</td><td>-0.3332074</td></tr>
    <tr><th scope=row>Gene4</th><td>3.070508</td><td>-0.7288912</td><td>5.168956</td><td>-1.0185754</td></tr>
</tbody>
</table>




<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">library(<span class="string">&quot;pheatmap&quot;</span>)</span><br><span class="line">pheatmap(mydata, scale = <span class="string">&quot;row&quot;</span>)</span><br></pre></td></tr></table></figure>

<pre><code>Warning message:
&quot;package &#39;pheatmap&#39; was built under R version 4.0.3&quot;
</code></pre>
<p><img src="img/output_94_1.png" alt="png"></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://example.com/2021/12/30/Hierarchical_clustering_in_R/" data-id="ckxswolvl0002qwmk516j4akm" data-title="" class="article-share-link">Share</a>
      
      
      
    </footer>
  </div>
  
    
<nav id="article-nav">
  
  
    <a href="/2021/12/30/GSEA%20instruction/" id="article-nav-older" class="article-nav-link-wrap">
      <strong class="article-nav-caption">Older</strong>
      <div class="article-nav-title"></div>
    </a>
  
</nav>

  
</article>


</section>
        
          <aside id="sidebar">
  
    

  
    

  
    
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Archives</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">December 2021</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Recent Posts</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2021/12/30/Hierarchical_clustering_in_R/">(no title)</a>
          </li>
        
          <li>
            <a href="/2021/12/30/GSEA%20instruction/">(no title)</a>
          </li>
        
          <li>
            <a href="/2021/12/30/Git_instructions/">(no title)</a>
          </li>
        
          <li>
            <a href="/2021/12/30/hello-world/">Hello World</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2021 shaoboliang<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.4.1.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>